{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "93992ae8",
   "metadata": {},
   "source": [
    "# How to train an ontology-based variational autoencoder (Ontix)\n",
    "\n",
    "In normal autoencoders latent dimensions are not explainable by design. To gain explainability and to incorporate biological information, a popular approach is to restrict the decoder of the autoencoder to match feature connectivity like an ontology. \n",
    "\n",
    "**IMPORTANT**\n",
    "\n",
    "> This tutorial only shows the specifics of the Ontix pipeline. If you're unfamilar with general concepts,  \n",
    "> we recommend to follow the `Getting Started - Vanillix` Tutorial first.\n",
    "## What You'll Learn\n",
    "\n",
    "In this notebook we will show two types of ontologies and how they can be used to train an explainable variational autoencoder `ontix`. \n",
    "The first is based on biological pathways of the Reactome database (left) and the second uses chromosomal location of genes (right) as a showcase. \n",
    "\n",
    "<img src=\"https://raw.githubusercontent.com/jan-forest/autoencodix/5dabc4a697cbba74d3f6144dc4b6d0fd6df2b624/images/ontix_scheme.svg\" alt=\"ontix-ontologies\" width=\"1200\"/>\n",
    "Youâ€™ll learn how to:\n",
    "\n",
    "1. **Initialize** the pipeline, and ontologies, then run the pipeline. <br> <br>\n",
    "2. Understand the Ontix sepecific **pipeline steps**. <br><br>\n",
    "3. Access the Ontox specific **results** (mus, sigma ,kl/mmd losses). <br><br>\n",
    "4. **Visualize** outputs effectively. <br><br>\n",
    "5. Apply **custom parameters**. <br><br>\n",
    "6. **Save, load, and reuse** a trained pipeline. <br><br>\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d1ab1d0d",
   "metadata": {},
   "source": [
    "## 01 Set-up Ontology and Initialize Pipeline\n",
    "\n",
    "The only thing you need to do to train an `ontix` is to provide text files for up to two ontology levels. The first ontology level is the mapping of your features (e.g. Gene ID) and an ontology level like subpathways or cytobands of chromosome. The second level is optional, but recommended, and is the mapping of the first ontology to a second level like top-level pathways or chromosome. \n",
    "\n",
    "The mapping should have the format:  \n",
    "Gene 1 `separator` Pathway1  \n",
    "Gene 2 `separator` Pathway1  \n",
    "Gene 3 `separator` Pathway2  <br><br>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ddc82c21",
   "metadata": {},
   "source": [
    "**Example 1: Set-up chromosomal ontology**:  \n",
    "\n",
    "From Ensembl via Biomart or any other sequence database you can get cytoband (karyotype) and chromosomal information for human genes like this:"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c84721ff",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "afb37d52",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.microsoft.datawrangler.viewer.v0+json": {
       "columns": [
        {
         "name": "index",
         "rawType": "int64",
         "type": "integer"
        },
        {
         "name": "Gene stable ID",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "Gene stable ID version",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "Karyotype band",
         "rawType": "object",
         "type": "unknown"
        },
        {
         "name": "Chromosome/scaffold name",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "Gene start (bp)",
         "rawType": "int64",
         "type": "integer"
        },
        {
         "name": "Gene end (bp)",
         "rawType": "int64",
         "type": "integer"
        },
        {
         "name": "HGNC symbol",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "NCBI gene (formerly Entrezgene) ID",
         "rawType": "int64",
         "type": "integer"
        }
       ],
       "ref": "71e3086e-ac86-4685-aa92-6c7ae6f56e5d",
       "rows": [
        [
         "0",
         "ENSG00000198888",
         "ENSG00000198888.2",
         null,
         "MT",
         "3307",
         "4262",
         "MT-ND1",
         "4535"
        ],
        [
         "1",
         "ENSG00000198763",
         "ENSG00000198763.3",
         null,
         "MT",
         "4470",
         "5511",
         "MT-ND2",
         "4536"
        ],
        [
         "2",
         "ENSG00000198804",
         "ENSG00000198804.2",
         null,
         "MT",
         "5904",
         "7445",
         "MT-CO1",
         "4512"
        ],
        [
         "3",
         "ENSG00000210151",
         "ENSG00000210151.2",
         null,
         "MT",
         "7446",
         "7514",
         "MT-TS1",
         "113219467"
        ],
        [
         "4",
         "ENSG00000198712",
         "ENSG00000198712.1",
         null,
         "MT",
         "7586",
         "8269",
         "MT-CO2",
         "4513"
        ]
       ],
       "shape": {
        "columns": 8,
        "rows": 5
       }
      },
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene stable ID</th>\n",
       "      <th>Gene stable ID version</th>\n",
       "      <th>Karyotype band</th>\n",
       "      <th>Chromosome/scaffold name</th>\n",
       "      <th>Gene start (bp)</th>\n",
       "      <th>Gene end (bp)</th>\n",
       "      <th>HGNC symbol</th>\n",
       "      <th>NCBI gene (formerly Entrezgene) ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000198888</td>\n",
       "      <td>ENSG00000198888.2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MT</td>\n",
       "      <td>3307</td>\n",
       "      <td>4262</td>\n",
       "      <td>MT-ND1</td>\n",
       "      <td>4535</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000198763</td>\n",
       "      <td>ENSG00000198763.3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MT</td>\n",
       "      <td>4470</td>\n",
       "      <td>5511</td>\n",
       "      <td>MT-ND2</td>\n",
       "      <td>4536</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000198804</td>\n",
       "      <td>ENSG00000198804.2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MT</td>\n",
       "      <td>5904</td>\n",
       "      <td>7445</td>\n",
       "      <td>MT-CO1</td>\n",
       "      <td>4512</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000210151</td>\n",
       "      <td>ENSG00000210151.2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MT</td>\n",
       "      <td>7446</td>\n",
       "      <td>7514</td>\n",
       "      <td>MT-TS1</td>\n",
       "      <td>113219467</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000198712</td>\n",
       "      <td>ENSG00000198712.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MT</td>\n",
       "      <td>7586</td>\n",
       "      <td>8269</td>\n",
       "      <td>MT-CO2</td>\n",
       "      <td>4513</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    Gene stable ID Gene stable ID version Karyotype band  \\\n",
       "0  ENSG00000198888      ENSG00000198888.2            NaN   \n",
       "1  ENSG00000198763      ENSG00000198763.3            NaN   \n",
       "2  ENSG00000198804      ENSG00000198804.2            NaN   \n",
       "3  ENSG00000210151      ENSG00000210151.2            NaN   \n",
       "4  ENSG00000198712      ENSG00000198712.1            NaN   \n",
       "\n",
       "  Chromosome/scaffold name  Gene start (bp)  Gene end (bp) HGNC symbol  \\\n",
       "0                       MT             3307           4262      MT-ND1   \n",
       "1                       MT             4470           5511      MT-ND2   \n",
       "2                       MT             5904           7445      MT-CO1   \n",
       "3                       MT             7446           7514      MT-TS1   \n",
       "4                       MT             7586           8269      MT-CO2   \n",
       "\n",
       "   NCBI gene (formerly Entrezgene) ID  \n",
       "0                                4535  \n",
       "1                                4536  \n",
       "2                                4512  \n",
       "3                           113219467  \n",
       "4                                4513  "
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "df_genes = pd.read_csv(\"genes_chromosomes.txt\", sep=\"\\t\")\n",
    "df_genes.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "80ead346",
   "metadata": {},
   "source": [
    "We have to solve some issues before we can use this as ontology.  \n",
    "(1) We only want chromosomes and not scaffolds  \n",
    "(2) Karyotype/cytoband should have identification of the chromosome in their name as identifier"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c7b414dd",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "This will be our chromosomes and latent dimensions in Ontix:\n",
      "['MT' 'Y' '21' '13' '18' '22' '20' 'X' '15' '14' '10' '9' '8' '16' '4' '5'\n",
      " '7' '6' '19' '12' '11' '3' '17' '2' '1']\n",
      "Latent dimension: 25\n"
     ]
    }
   ],
   "source": [
    "df_genes = df_genes.loc[\n",
    "    df_genes[\"Chromosome/scaffold name\"].str.len() < 3\n",
    "]  ## get rid of scaffolds and keep only chromosomes\n",
    "df_genes.loc[df_genes[\"Chromosome/scaffold name\"] == \"MT\", \"Karyotype band\"] = (\n",
    "    \"MT\"  ## create missing karyotype for mito genes\n",
    ")\n",
    "print(\"This will be our chromosomes and latent dimensions in Ontix:\")\n",
    "print(df_genes[\"Chromosome/scaffold name\"].unique())\n",
    "print(f\"Latent dimension: {len(df_genes['Chromosome/scaffold name'].unique())}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "03bd6338",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "This will be our hidden layer in the sparse decoder:\n",
      "['MT:MT' 'Y:p11.2' 'Y:q11.223' 'Y:q11.221' 'Y:q11.222' 'Y:q11.23'\n",
      " 'Y:p11.31' 'Y:p11.32' 'Y:q12' '21:p12' '21:q21.1' '21:q21.2' '21:p11.2'\n",
      " '13:q12.12' '21:q21.3' '21:q22.11' '13:q12.3' '13:q14.12' '13:q14.2'\n",
      " '21:q22.12']\n",
      "Hidden layer dim: 817\n"
     ]
    }
   ],
   "source": [
    "# Combine Chromosome name and cytoband\n",
    "df_genes = df_genes.copy()\n",
    "df_genes.loc[:, \"Chr_and_karyotype\"] = df_genes.loc[\n",
    "    :, [\"Chromosome/scaffold name\", \"Karyotype band\"]\n",
    "].apply(lambda x: \":\".join(x.values.tolist()), axis=1)\n",
    "print(\"This will be our hidden layer in the sparse decoder:\")\n",
    "print(df_genes[\"Chr_and_karyotype\"].unique()[0:20])\n",
    "print(f\"Hidden layer dim: {len(df_genes['Chr_and_karyotype'].unique())}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "08523ca7",
   "metadata": {},
   "source": [
    "Now we can save this as files in the correct format for the two levels:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "efc773d0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Changed to: /Users/maximilianjoas/development/autoencodix_package\n"
     ]
    },
    {
     "ename": "TypeError",
     "evalue": "join() argument must be str, bytes, or os.PathLike object, not 'tuple'",
     "output_type": "error",
     "traceback": [
      "\u001b[31m---------------------------------------------------------------------------\u001b[39m",
      "\u001b[31mTypeError\u001b[39m                                 Traceback (most recent call last)",
      "\u001b[36mCell\u001b[39m\u001b[36m \u001b[39m\u001b[32mIn[8]\u001b[39m\u001b[32m, line 28\u001b[39m\n\u001b[32m     17\u001b[39m ont_file2 = \u001b[33m\"\u001b[39m\u001b[33mdata/raw/chromosome_ont_lvl2.txt\u001b[39m\u001b[33m\"\u001b[39m,\n\u001b[32m     20\u001b[39m \u001b[38;5;66;03m# Level 1\u001b[39;00m\n\u001b[32m     21\u001b[39m df_genes[\n\u001b[32m     22\u001b[39m     [\n\u001b[32m     23\u001b[39m         \u001b[33m\"\u001b[39m\u001b[33mNCBI gene (formerly Entrezgene) ID\u001b[39m\u001b[33m\"\u001b[39m,\n\u001b[32m     24\u001b[39m         \u001b[33m\"\u001b[39m\u001b[33mChr_and_karyotype\u001b[39m\u001b[33m\"\u001b[39m,\n\u001b[32m     25\u001b[39m     ]  \u001b[38;5;66;03m# Level 1: Feature (gene) to hidden layer (cytoband)\u001b[39;00m\n\u001b[32m     26\u001b[39m ].drop_duplicates(  \u001b[38;5;66;03m# Chromosomal ontology must be unique\u001b[39;00m\n\u001b[32m     27\u001b[39m ).to_csv(\n\u001b[32m---> \u001b[39m\u001b[32m28\u001b[39m     \u001b[43mos\u001b[49m\u001b[43m.\u001b[49m\u001b[43mpath\u001b[49m\u001b[43m.\u001b[49m\u001b[43mjoin\u001b[49m\u001b[43m(\u001b[49m\u001b[43mdata_root\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mont_file1\u001b[49m\u001b[43m)\u001b[49m,\n\u001b[32m     29\u001b[39m     sep=\u001b[33m\"\u001b[39m\u001b[38;5;130;01m\\t\u001b[39;00m\u001b[33m\"\u001b[39m,\n\u001b[32m     30\u001b[39m     header=\u001b[38;5;28;01mFalse\u001b[39;00m,\n\u001b[32m     31\u001b[39m     index=\u001b[38;5;28;01mFalse\u001b[39;00m,\n\u001b[32m     32\u001b[39m )\n\u001b[32m     34\u001b[39m \u001b[38;5;66;03m# Level 2\u001b[39;00m\n\u001b[32m     35\u001b[39m df_genes[\n\u001b[32m     36\u001b[39m     [\n\u001b[32m     37\u001b[39m         \u001b[33m\"\u001b[39m\u001b[33mChr_and_karyotype\u001b[39m\u001b[33m\"\u001b[39m,\n\u001b[32m   (...)\u001b[39m\u001b[32m     46\u001b[39m     index=\u001b[38;5;28;01mFalse\u001b[39;00m,\n\u001b[32m     47\u001b[39m )\n",
      "\u001b[36mFile \u001b[39m\u001b[32m<frozen posixpath>:90\u001b[39m, in \u001b[36mjoin\u001b[39m\u001b[34m(a, *p)\u001b[39m\n",
      "\u001b[36mFile \u001b[39m\u001b[32m<frozen genericpath>:164\u001b[39m, in \u001b[36m_check_arg_types\u001b[39m\u001b[34m(funcname, *args)\u001b[39m\n",
      "\u001b[31mTypeError\u001b[39m: join() argument must be str, bytes, or os.PathLike object, not 'tuple'"
     ]
    }
   ],
   "source": [
    "import os\n",
    "\n",
    "p = os.getcwd()\n",
    "d = \"autoencodix_package\"\n",
    "if d not in p:\n",
    "    raise FileNotFoundError(f\"'{d}' not found in path: {p}\")\n",
    "os.chdir(os.sep.join(p.split(os.sep)[: p.split(os.sep).index(d) + 1]))\n",
    "print(f\"Changed to: {os.getcwd()}\")\n",
    "# ---------------------------------------------------------------------\n",
    "# Paths\n",
    "# ---------------------------------------------------------------------\n",
    "data_root = \"data/raw\"\n",
    "rna_file = \"combined_rnaseq_formatted.parquet\"\n",
    "meth_file = \"combined_meth_formatted.parquet\"\n",
    "clin_file = \"combined_clin_formatted.parquet\"\n",
    "ont_file1 = \"chromosome_ont_lvl1_ncbi.txt\",\n",
    "ont_file2 = \"data/raw/chromosome_ont_lvl2.txt\",\n",
    "\n",
    "\n",
    "# Level 1\n",
    "df_genes[\n",
    "    [\n",
    "        \"NCBI gene (formerly Entrezgene) ID\",\n",
    "        \"Chr_and_karyotype\",\n",
    "    ]  # Level 1: Feature (gene) to hidden layer (cytoband)\n",
    "].drop_duplicates(  # Chromosomal ontology must be unique\n",
    ").to_csv(\n",
    "    os.path.join(data_root, ont_file1),\n",
    "    sep=\"\\t\",\n",
    "    header=False,\n",
    "    index=False,\n",
    ")\n",
    "\n",
    "# Level 2\n",
    "df_genes[\n",
    "    [\n",
    "        \"Chr_and_karyotype\",\n",
    "        \"Chromosome/scaffold name\",\n",
    "    ]  # Level 2: hidden layer (cytoband) to latent dimension (chromosome)\n",
    "].drop_duplicates(  # Chromosomal ontology must be unique\n",
    ").to_csv(\n",
    "\n",
    "    os.path.join(data_root, ont_file2),\n",
    "    sep=\"\\t\",\n",
    "    header=False,\n",
    "    index=False,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "9e46af96",
   "metadata": {},
   "outputs": [
    {
     "ename": "ValidationError",
     "evalue": "1 validation error for OntixConfig\n  Value error, Invalid scaling 'NOTSET' for modality 'RNA'. OntixConfig only permits ['MINMAX', 'NONE']. [type=value_error, input_value={'data_config': DataConfi...0.2, 'valid_ratio': 0.1}, input_type=dict]\n    For further information visit https://errors.pydantic.dev/2.12/v/value_error",
     "output_type": "error",
     "traceback": [
      "\u001b[31m---------------------------------------------------------------------------\u001b[39m",
      "\u001b[31mValidationError\u001b[39m                           Traceback (most recent call last)",
      "\u001b[36mCell\u001b[39m\u001b[36m \u001b[39m\u001b[32mIn[12]\u001b[39m\u001b[32m, line 57\u001b[39m\n\u001b[32m     37\u001b[39m data_config = DataConfig(\n\u001b[32m     38\u001b[39m     data_info={\n\u001b[32m     39\u001b[39m         \u001b[33m\"\u001b[39m\u001b[33mRNA\u001b[39m\u001b[33m\"\u001b[39m: rna_info,\n\u001b[32m   (...)\u001b[39m\u001b[32m     51\u001b[39m     ],\n\u001b[32m     52\u001b[39m )\n\u001b[32m     54\u001b[39m \u001b[38;5;66;03m# ---------------------------------------------------------------------\u001b[39;00m\n\u001b[32m     55\u001b[39m \u001b[38;5;66;03m# Define the full DefaultConfig (roughly equivalent to old cfg)\u001b[39;00m\n\u001b[32m     56\u001b[39m \u001b[38;5;66;03m# ---------------------------------------------------------------------\u001b[39;00m\n\u001b[32m---> \u001b[39m\u001b[32m57\u001b[39m ontix_config = \u001b[43mOntixConfig\u001b[49m\u001b[43m(\u001b[49m\n\u001b[32m     58\u001b[39m \u001b[43m    \u001b[49m\u001b[43mdata_config\u001b[49m\u001b[43m=\u001b[49m\u001b[43mdata_config\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m     59\u001b[39m \u001b[43m    \u001b[49m\u001b[43mreproducible\u001b[49m\u001b[43m=\u001b[49m\u001b[38;5;28;43;01mTrue\u001b[39;49;00m\u001b[43m,\u001b[49m\n\u001b[32m     60\u001b[39m \u001b[43m    \u001b[49m\u001b[43mglobal_seed\u001b[49m\u001b[43m=\u001b[49m\u001b[32;43m42\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[32m     61\u001b[39m \u001b[43m    \u001b[49m\u001b[43mepochs\u001b[49m\u001b[43m=\u001b[49m\u001b[32;43m500\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[32m     62\u001b[39m \u001b[43m    \u001b[49m\u001b[43mlearning_rate\u001b[49m\u001b[43m=\u001b[49m\u001b[32;43m0.0005\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[32m     63\u001b[39m \u001b[43m    \u001b[49m\u001b[43mbatch_size\u001b[49m\u001b[43m=\u001b[49m\u001b[32;43m128\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[32m     64\u001b[39m \u001b[43m    \u001b[49m\u001b[43mdrop_p\u001b[49m\u001b[43m=\u001b[49m\u001b[32;43m0.3\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[32m     65\u001b[39m \u001b[43m    \u001b[49m\u001b[43mk_filter\u001b[49m\u001b[43m=\u001b[49m\u001b[32;43m1000\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[32m     66\u001b[39m \u001b[43m    \u001b[49m\u001b[43mlatent_dim\u001b[49m\u001b[43m=\u001b[49m\u001b[32;43m6\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\n\u001b[32m     67\u001b[39m \u001b[43m    \u001b[49m\u001b[43mreconstruction_loss\u001b[49m\u001b[43m=\u001b[49m\u001b[33;43m\"\u001b[39;49m\u001b[33;43mmse\u001b[39;49m\u001b[33;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[32m     68\u001b[39m \u001b[43m    \u001b[49m\u001b[43mdefault_vae_loss\u001b[49m\u001b[43m=\u001b[49m\u001b[33;43m\"\u001b[39;49m\u001b[33;43mkl\u001b[39;49m\u001b[33;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[32m     69\u001b[39m \u001b[43m    \u001b[49m\u001b[43mbeta\u001b[49m\u001b[43m=\u001b[49m\u001b[32;43m0.5\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[32m     70\u001b[39m \u001b[43m    \u001b[49m\u001b[43msave_memory\u001b[49m\u001b[43m=\u001b[49m\u001b[38;5;28;43;01mTrue\u001b[39;49;00m\u001b[43m,\u001b[49m\n\u001b[32m     71\u001b[39m \u001b[43m    \u001b[49m\u001b[43mscaling\u001b[49m\u001b[43m=\u001b[49m\u001b[33;43m\"\u001b[39;49m\u001b[33;43mMINMAX\u001b[39;49m\u001b[33;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[32m     72\u001b[39m \u001b[43m    \u001b[49m\u001b[43mdevice\u001b[49m\u001b[43m=\u001b[49m\u001b[33;43m\"\u001b[39;49m\u001b[33;43mauto\u001b[39;49m\u001b[33;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[32m     73\u001b[39m \u001b[43m    \u001b[49m\u001b[43mtrain_ratio\u001b[49m\u001b[43m=\u001b[49m\u001b[32;43m0.7\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[32m     74\u001b[39m \u001b[43m    \u001b[49m\u001b[43mtest_ratio\u001b[49m\u001b[43m=\u001b[49m\u001b[32;43m0.2\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[32m     75\u001b[39m \u001b[43m    \u001b[49m\u001b[43mvalid_ratio\u001b[49m\u001b[43m=\u001b[49m\u001b[32;43m0.1\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[32m     76\u001b[39m \u001b[43m)\u001b[49m\n\u001b[32m     78\u001b[39m \u001b[38;5;66;03m# ---------------------------------------------------------------------\u001b[39;00m\n\u001b[32m     79\u001b[39m \u001b[38;5;66;03m# Now pass into your Ontix object\u001b[39;00m\n\u001b[32m     80\u001b[39m \u001b[38;5;66;03m# ---------------------------------------------------------------------\u001b[39;00m\n\u001b[32m     82\u001b[39m ont_files = [ont_file1, ont_file2]\n",
      "\u001b[36mFile \u001b[39m\u001b[32m~/development/autoencodix_package/.venv/lib/python3.12/site-packages/pydantic/main.py:250\u001b[39m, in \u001b[36mBaseModel.__init__\u001b[39m\u001b[34m(self, **data)\u001b[39m\n\u001b[32m    248\u001b[39m \u001b[38;5;66;03m# `__tracebackhide__` tells pytest and some other tools to omit this function from tracebacks\u001b[39;00m\n\u001b[32m    249\u001b[39m __tracebackhide__ = \u001b[38;5;28;01mTrue\u001b[39;00m\n\u001b[32m--> \u001b[39m\u001b[32m250\u001b[39m validated_self = \u001b[38;5;28;43mself\u001b[39;49m\u001b[43m.\u001b[49m\u001b[43m__pydantic_validator__\u001b[49m\u001b[43m.\u001b[49m\u001b[43mvalidate_python\u001b[49m\u001b[43m(\u001b[49m\u001b[43mdata\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mself_instance\u001b[49m\u001b[43m=\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[43m)\u001b[49m\n\u001b[32m    251\u001b[39m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mself\u001b[39m \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m validated_self:\n\u001b[32m    252\u001b[39m     warnings.warn(\n\u001b[32m    253\u001b[39m         \u001b[33m'\u001b[39m\u001b[33mA custom validator is returning a value other than `self`.\u001b[39m\u001b[38;5;130;01m\\n\u001b[39;00m\u001b[33m'\u001b[39m\n\u001b[32m    254\u001b[39m         \u001b[33m\"\u001b[39m\u001b[33mReturning anything other than `self` from a top level model validator isn\u001b[39m\u001b[33m'\u001b[39m\u001b[33mt supported when validating via `__init__`.\u001b[39m\u001b[38;5;130;01m\\n\u001b[39;00m\u001b[33m\"\u001b[39m\n\u001b[32m    255\u001b[39m         \u001b[33m'\u001b[39m\u001b[33mSee the `model_validator` docs (https://docs.pydantic.dev/latest/concepts/validators/#model-validators) for more details.\u001b[39m\u001b[33m'\u001b[39m,\n\u001b[32m    256\u001b[39m         stacklevel=\u001b[32m2\u001b[39m,\n\u001b[32m    257\u001b[39m     )\n",
      "\u001b[31mValidationError\u001b[39m: 1 validation error for OntixConfig\n  Value error, Invalid scaling 'NOTSET' for modality 'RNA'. OntixConfig only permits ['MINMAX', 'NONE']. [type=value_error, input_value={'data_config': DataConfi...0.2, 'valid_ratio': 0.1}, input_type=dict]\n    For further information visit https://errors.pydantic.dev/2.12/v/value_error"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import autoencodix as acx\n",
    "from autoencodix.configs.default_config import DataConfig, DataInfo, DataCase\n",
    "from autoencodix.configs import OntixConfig\n",
    "\n",
    "# ---------------------------------------------------------------------\n",
    "# Paths\n",
    "# ---------------------------------------------------------------------\n",
    "data_root = \"data/raw\"\n",
    "rna_file = \"combined_rnaseq_formatted.parquet\"\n",
    "meth_file = \"combined_meth_formatted.parquet\"\n",
    "clin_file = \"combined_clin_formatted.parquet\"\n",
    "\n",
    "# ---------------------------------------------------------------------\n",
    "# Define individual data modalities\n",
    "# ---------------------------------------------------------------------\n",
    "rna_info = DataInfo(\n",
    "    file_path=os.path.join(data_root, rna_file),\n",
    "    data_type=\"NUMERIC\",\n",
    "    filtering=\"VAR\",\n",
    ")\n",
    "\n",
    "meth_info = DataInfo(\n",
    "    file_path=os.path.join(data_root, meth_file),\n",
    "    data_type=\"NUMERIC\",\n",
    "    filtering=\"VAR\",\n",
    ")\n",
    "\n",
    "anno_info = DataInfo(\n",
    "    file_path=os.path.join(data_root, clin_file),\n",
    "    data_type=\"ANNOTATION\",\n",
    ")\n",
    "\n",
    "# ---------------------------------------------------------------------\n",
    "# Combine into DataConfig\n",
    "# ---------------------------------------------------------------------\n",
    "data_config = DataConfig(\n",
    "    data_info={\n",
    "        \"RNA\": rna_info,\n",
    "        \"METH\": meth_info,\n",
    "        \"ANNO\": anno_info,\n",
    "    },\n",
    "    annotation_columns=[\n",
    "        \"CANCER_TYPE\",\n",
    "        \"CANCER_TYPE_ACRONYM\",\n",
    "        \"TMB_NONSYNONYMOUS\",\n",
    "        \"AGE\",\n",
    "        \"OS_STATUS\",\n",
    "        \"GRADE\",\n",
    "        \"SEX\",\n",
    "    ],\n",
    ")\n",
    "\n",
    "# ---------------------------------------------------------------------\n",
    "# Define the full DefaultConfig (roughly equivalent to old cfg)\n",
    "# ---------------------------------------------------------------------\n",
    "ontix_config = OntixConfig(\n",
    "    data_config=data_config,\n",
    "    reproducible=True,\n",
    "    global_seed=42,\n",
    "    epochs=500,\n",
    "    learning_rate=0.0005,\n",
    "    batch_size=128,\n",
    "    drop_p=0.3,\n",
    "    k_filter=1000,\n",
    "    latent_dim=6, \n",
    "    reconstruction_loss=\"mse\",\n",
    "    default_vae_loss=\"kl\",\n",
    "    beta=0.5,\n",
    "    save_memory=True,\n",
    "    scaling=\"MINMAX\",\n",
    "    device=\"auto\",\n",
    "    train_ratio=0.7,\n",
    "    test_ratio=0.2,\n",
    "    valid_ratio=0.1,\n",
    ")\n",
    "\n",
    "# ---------------------------------------------------------------------\n",
    "# Now pass into your Ontix object\n",
    "# ---------------------------------------------------------------------\n",
    "\n",
    "ont_files = [ont_file1, ont_file2]\n",
    "ontix = acx.Ontix(\n",
    "    ontologies=ont_files,\n",
    "    config=ontix_config,\n",
    ")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "86ef7e13",
   "metadata": {},
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "autoencodix",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
