{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "df474bf2",
   "metadata": {},
   "source": [
    "# How To Use XModalix\n",
    "\n",
    "X-Modalix is our implementation of a cross-modal autoencoder.  \n",
    "This tutorial follows the structure of `Getting Started - Vanillix`, but is less extensive because our pipeline works similarly across different architectures. Here, we focus only on X-Modalix specifics.\n",
    "\n",
    "**AUTOENCODIX** supports far more functionality than shown here, so weâ€™ll also point to advanced tutorials where relevant.  \n",
    "\n",
    "**IMPORTANT**\n",
    "\n",
    "> This tutorial only shows the specifics of the XModalix pipeline. If you're unfamiliar with general concepts,  \n",
    "> we recommend following the `Getting Started - Vanillix` tutorial first.\n",
    "\n",
    "## XModalix Theory\n",
    "\n",
    "The core idea of a cross-modal autoencoder is to align the latent spaces of different data modalities.  \n",
    "Once the latent distributions are aligned, we can translate between modalities by feeding data from one modality into its encoder and then passing the latent space to the decoder of the target modality.\n",
    "\n",
    "The underlying models are separate VAEsâ€”one for each data modalityâ€”which are trained in parallel. Their latent spaces are closely aligned so that modalities are hard to discriminate, but sample variation is preserved.  \n",
    "To achieve this alignment, the VAE loss function is extended with the following terms:\n",
    "\n",
    "- *Adversarial* loss term: In parallel with the VAEs, a third neural network (latent space classifier) is trained to discriminate the embeddings of different modalities. Analogous to a GAN, the inverse of this classifier loss is added to the VAE loss to enforce latent space alignment.  \n",
    "\n",
    "- *Paired* loss term: Alignment is further enforced by minimizing the distance between samples with paired measurements across modalities. While unpaired samples are allowed, we recommend that most samples are paired (present in all modalities).  \n",
    "\n",
    "- *Class-based* loss term: This semi-supervised term minimizes distances among samples of the same class. Class membership is defined in the annotation data (e.g., cancer type, cell type). In each iteration, the class mean in the latent space is computed, and distances from individual samples to the class mean are minimized. See our `Preprint [TODO]` for more details.  \n",
    "\n",
    "<img src=\"https://raw.githubusercontent.com/jan-forest/autoencodix/729f3e8172a722e7c3681298719d9d4b59e313d9/images/xmodalix_scheme.svg\" alt=\"xmodalix_scheme\" width=\"1200\"/>\n",
    "\n",
    "## What You'll Learn\n",
    "\n",
    "As a showcase for data modality translation with X-Modalix, we will use cancer gene expression from TCGA in combination with handwritten digits from the [MNIST dataset](https://keras.io/api/datasets/mnist/).  \n",
    "Our goal is to translate gene expression signatures of five selected cancer subtypes into images of digits, with each cancer subtype class mapped to a digit between 0â€“4.  \n",
    "\n",
    "In practice, these images could be histopathological images or any other data modality.  \n",
    "Before we show data preparation and X-Modalix training, we provide background on cross-modal VAEs as proposed by [Yang & Uhler](https://arxiv.org/abs/1902.03515).  \n",
    "\n",
    "While following this showcase, you will learn how to:\n",
    "\n",
    "1. **Initialize** the pipeline and run it. <br><br>\n",
    "2. Understand the X-Modalix-specific **pipeline steps** (paired vs. unpaired data). <br><br>\n",
    "3. Access X-Modalix-specific **results** (sub-results for modality autoencoders). <br><br>\n",
    "4. **Visualize** and **evaluate** the outputs. <br><br>\n",
    "5. Apply **custom parameters**. <br><br>\n",
    "6. **Save, load, and reuse** a trained pipeline. <br><br>\n",
    "\n",
    "Letâ€™s get started! ðŸš€\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6d2465f4",
   "metadata": {},
   "source": [
    "## 1) Initialize and Run XModalix\n",
    "\n",
    "In this example, we read our input data from files.  \n",
    "The file locations are defined in the config via a section called `DataConfig` (see details in the Python code below).  \n",
    "\n",
    "We highlight a few custom config parameters for XModalix. For a deeper dive into the config object, see [1]:\n",
    "\n",
    "- `pretrain_epochs`: Before training the full X-Modalix, the sub-VAEs of each modality can be pretrained. This parameter can be set globally for all modalities or individually per modality and specifies the number of pretraining epochs.  \n",
    "- `gamma`: Hyperparameter to weight the adversarial loss term.  \n",
    "- `delta_pair`: Hyperparameter to weight the paired loss term.  \n",
    "- `delta_class`: Hyperparameter to weight the class-based loss term.  \n",
    "\n",
    "### Extra 1: Getting Tutorial Data\n",
    "\n",
    "You can run the following bash commands to download the data and set up the correct folder structure.  \n",
    "**Assumption:** you are in the root of `autoencodix_package`.\n",
    "\n",
    "```bash\n",
    "mkdir -p data\n",
    "cd data\n",
    "wget \"https://cloud.scadsai.uni-leipzig.de/index.php/s/bq64MaQyZGZfN64/download/XModalix-Tut-data.zip\"\n",
    "unzip XModalix-Tut-data.zip -d XModalix-Tut-data\n",
    "```\n",
    "\n",
    "### Extra 2: Get correct path\n",
    "We assume you are in the root of the package. The following code ensures that the correct paths are used.\n",
    "[1] Tutorials/DeepDives/ConfigTutorial.ipynb"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "d4da076c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Changed to: /Users/maximilianjoas/development/autoencodix_package\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "\n",
    "p = os.getcwd()\n",
    "d = \"autoencodix_package\"\n",
    "if d not in p:\n",
    "    raise FileNotFoundError(f\"'{d}' not found in path: {p}\")\n",
    "os.chdir(os.sep.join(p.split(os.sep)[: p.split(os.sep).index(d) + 1]))\n",
    "print(f\"Changed to: {os.getcwd()}\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ad61c1a7",
   "metadata": {},
   "source": [
    "### Extra 3: Data Assumptions\n",
    "\n",
    "X-Modalix supports two special cases:  \n",
    "(a) image data, and  \n",
    "(b) unpaired data.  \n",
    "\n",
    "This introduces specific requirements for the input data. In the config object, the `DataConfig` must be populated with the relevant information. Example code:\n",
    "\n",
    "```python\n",
    "imganno_file = os.path.join(\"data/XModalix-Tut-data/tcga_mappings.txt\")\n",
    "rna_file = os.path.join(\"data/XModalix-Tut-data/combined_rnaseq_formatted.parquet\")\n",
    "img_root = os.path.join(\"data/XModalix-Tut-data/images/tcga_fake\")\n",
    "clin_file = os.path.join(\"data/XModalix-Tut-data/combined_clin_formatted.parquet\")\n",
    "\n",
    "dc = DataConfig(\n",
    "    data_info={\n",
    "        \"img\": DataInfo(\n",
    "            file_path=img_root,\n",
    "            data_type=\"IMG\",\n",
    "            scaling=\"MINMAX\",\n",
    "            translate_direction=\"to\",\n",
    "            pretrain_epochs=3,\n",
    "            # extra_anno_file=imganno_file,\n",
    "        ),\n",
    "        \"rna\": DataInfo(\n",
    "            file_path=rna_file,\n",
    "            data_type=\"NUMERIC\",\n",
    "            scaling=\"MINMAX\",\n",
    "            translate_direction=\"from\",\n",
    "        ),\n",
    "        \"anno\": DataInfo(\n",
    "            file_path=clin_file,\n",
    "            data_type=\"ANNOTATION\",\n",
    "            sep=\"\\t\"\n",
    "        ),\n",
    "    }\n",
    ")\n",
    "```\n",
    "**Image Data Requirements**\n",
    "\n",
    "- Specify the root folder where the images are located. The preprocessor will automatically read all images and convert them into tensors.  \n",
    "- Supported image extensions (case-insensitive): `\".jpg\"`, `\".jpeg\"`, `\".png\"`, `\".tif\"`, `\".tiff\"`.  \n",
    "- A mapping between each image filename (without directories) and a `sample_id` is required, along with any additional metadata.  \n",
    "\n",
    "To provide this mapping, you must supply a **global annotation file** that contains all `sample_id`s and metadata columns, including one column with the image paths.  \n",
    "- Use the config parameter `img_path_col` to define the name of the column containing image paths (default: `\"img_paths\"`).  \n",
    "- See your `clin_file` for an example.\n",
    "\n",
    "**Translation Direction**\n",
    "\n",
    "- X-Modalix trains a shared latent alignment, enabling translation between all modalities.  \n",
    "- To keep results consistent, users should define a `translate_direction` in the config.  \n",
    "- If you want to change the direction or translate a different pair, you can specify this in the `predict` step (see [Section 3](#3-access-xmodalix-results)).\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "7ed3befc",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "First five image paths:\n",
      "['2331_label_0.png', '2332_label_0.png', '2333_label_0.png', '2334_label_0.png', '2335_label_0.png']\n",
      "\n",
      "\n"
     ]
    },
    {
     "data": {
      "application/vnd.microsoft.datawrangler.viewer.v0+json": {
       "columns": [
        {
         "name": "SAMPLE_ID",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "ONCOTREE_CODE",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "CANCER_TYPE",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "CANCER_TYPE_DETAILED",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "TUMOR_TYPE",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "GRADE",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "TISSUE_PROSPECTIVE_COLLECTION_INDICATOR",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "TISSUE_RETROSPECTIVE_COLLECTION_INDICATOR",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "TISSUE_SOURCE_SITE_CODE",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "TUMOR_TISSUE_SITE",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "ANEUPLOIDY_SCORE",
         "rawType": "float64",
         "type": "float"
        },
        {
         "name": "SAMPLE_TYPE",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "MSI_SCORE_MANTIS",
         "rawType": "float64",
         "type": "float"
        },
        {
         "name": "MSI_SENSOR_SCORE",
         "rawType": "float64",
         "type": "float"
        },
        {
         "name": "SOMATIC_STATUS",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "TMB_NONSYNONYMOUS",
         "rawType": "float64",
         "type": "float"
        },
        {
         "name": "TISSUE_SOURCE_SITE",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "TBL_SCORE",
         "rawType": "float64",
         "type": "float"
        },
        {
         "name": "SUBTYPE",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "CANCER_TYPE_ACRONYM",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "OTHER_PATIENT_ID",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "AGE",
         "rawType": "float64",
         "type": "float"
        },
        {
         "name": "SEX",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "AJCC_PATHOLOGIC_TUMOR_STAGE",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "AJCC_STAGING_EDITION",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "DAYS_LAST_FOLLOWUP",
         "rawType": "float64",
         "type": "float"
        },
        {
         "name": "DAYS_TO_BIRTH",
         "rawType": "float64",
         "type": "float"
        },
        {
         "name": "DAYS_TO_INITIAL_PATHOLOGIC_DIAGNOSIS",
         "rawType": "float64",
         "type": "float"
        },
        {
         "name": "ETHNICITY",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "FORM_COMPLETION_DATE",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "HISTORY_NEOADJUVANT_TRTYN",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "ICD_10",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "ICD_O_3_HISTOLOGY",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "ICD_O_3_SITE",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "INFORMED_CONSENT_VERIFIED",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "NEW_TUMOR_EVENT_AFTER_INITIAL_TREATMENT",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "PATH_M_STAGE",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "PATH_N_STAGE",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "PATH_T_STAGE",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "PERSON_NEOPLASM_CANCER_STATUS",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "PRIMARY_LYMPH_NODE_PRESENTATION_ASSESSMENT",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "PRIOR_DX",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "RACE",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "RADIATION_THERAPY",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "WEIGHT",
         "rawType": "float64",
         "type": "float"
        },
        {
         "name": "IN_PANCANPATHWAYS_FREEZE",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "OS_STATUS",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "OS_MONTHS",
         "rawType": "float64",
         "type": "float"
        },
        {
         "name": "DSS_STATUS",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "DSS_MONTHS",
         "rawType": "float64",
         "type": "float"
        },
        {
         "name": "DFS_STATUS",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "DFS_MONTHS",
         "rawType": "float64",
         "type": "float"
        },
        {
         "name": "PFS_STATUS",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "PFS_MONTHS",
         "rawType": "float64",
         "type": "float"
        },
        {
         "name": "GENETIC_ANCESTRY_LABEL",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "img_paths",
         "rawType": "object",
         "type": "string"
        }
       ],
       "ref": "b99ddc55-7da8-405a-bce4-6c7d5aabdf8f",
       "rows": [
        [
         "TCGA-3C-AAAU-01",
         "ILC",
         "Breast Cancer",
         "Breast Invasive Lobular Carcinoma",
         "Infiltrating Lobular Carcinoma",
         "unknown",
         "No",
         "Yes",
         "3C",
         "Breast",
         "19.0",
         "Primary",
         "0.3319",
         "0.55",
         "Matched",
         "0.8",
         "Columbia University",
         "205.0",
         "BRCA_LumA",
         "BRCA",
         "6E7D5EC6-A469-467C-B748-237353C23416",
         "55.0",
         "Female",
         "STAGE X",
         "6TH",
         "4047.0",
         "-20211.0",
         "0.0",
         "Not Hispanic Or Latino",
         "1/13/14",
         "No",
         "C50.9",
         "8520/3",
         "C50.9",
         "Yes",
         "No",
         "MX",
         "NX",
         "TX",
         "With Tumor",
         "Yes",
         "No",
         "White",
         "No",
         null,
         "Yes",
         "0:LIVING",
         "133.0505967",
         "0:ALIVE OR DEAD TUMOR FREE",
         "133.0505967",
         "1:Recurred/Progressed",
         "59.44044449",
         "1:PROGRESSION",
         "59.44044449",
         "EUR",
         "2331_label_0.png"
        ],
        [
         "TCGA-3C-AALI-01",
         "IDC",
         "Breast Cancer",
         "Breast Invasive Ductal Carcinoma",
         "Infiltrating Ductal Carcinoma",
         "unknown",
         "No",
         "Yes",
         "3C",
         "Breast",
         "22.0",
         "Primary",
         "0.3449",
         "0.74",
         "Matched",
         "15.26666667",
         "Columbia University",
         "190.0",
         "BRCA_Her2",
         "BRCA",
         "55262FCB-1B01-4480-B322-36570430C917",
         "50.0",
         "Female",
         "STAGE IIB",
         "6TH",
         "4005.0",
         "-18538.0",
         "0.0",
         "Not Hispanic Or Latino",
         "7/28/14",
         "No",
         "C50.9",
         "8500/3",
         "C50.9",
         "Yes",
         "No",
         "M0",
         "N1A",
         "T2",
         "Tumor Free",
         "Yes",
         "No",
         "Black or African American",
         "Yes",
         null,
         "Yes",
         "0:LIVING",
         "131.6697899",
         "0:ALIVE OR DEAD TUMOR FREE",
         "131.6697899",
         "0:DiseaseFree",
         "131.6697899",
         "0:CENSORED",
         "131.6697899",
         "AFR",
         "2332_label_0.png"
        ],
        [
         "TCGA-3C-AALJ-01",
         "IDC",
         "Breast Cancer",
         "Breast Invasive Ductal Carcinoma",
         "Infiltrating Ductal Carcinoma",
         "unknown",
         "No",
         "Yes",
         "3C",
         "Breast",
         "13.0",
         "Primary",
         "0.3266",
         "0.31",
         "Matched",
         "0.933333333",
         "Columbia University",
         "365.0",
         "BRCA_LumB",
         "BRCA",
         "427D0648-3F77-4FFC-B52C-89855426D647",
         "62.0",
         "Female",
         "STAGE IIB",
         "7TH",
         "1474.0",
         "-22848.0",
         "0.0",
         "Not Hispanic Or Latino",
         "7/28/14",
         "No",
         "C50.9",
         "8500/3",
         "C50.9",
         "Yes",
         "No",
         "M0",
         "N1A",
         "T2",
         "Tumor Free",
         "Yes",
         "No",
         "Black or African American",
         "No",
         null,
         "Yes",
         "0:LIVING",
         "48.45974291",
         "0:ALIVE OR DEAD TUMOR FREE",
         "48.45974291",
         "0:DiseaseFree",
         "48.45974291",
         "0:CENSORED",
         "48.45974291",
         "AFR_ADMIX",
         "2333_label_0.png"
        ],
        [
         "TCGA-3C-AALK-01",
         "IDC",
         "Breast Cancer",
         "Breast Invasive Ductal Carcinoma",
         "Infiltrating Ductal Carcinoma",
         "unknown",
         "No",
         "Yes",
         "3C",
         "Breast",
         "4.0",
         "Primary",
         "0.3218",
         "0.03",
         "Matched",
         "1.5",
         "Columbia University",
         "25.0",
         "BRCA_LumA",
         "BRCA",
         "C31900A4-5DCD-4022-97AC-638E86E889E4",
         "52.0",
         "Female",
         "STAGE IA",
         "7TH",
         "1448.0",
         "-19074.0",
         "0.0",
         "Not Hispanic Or Latino",
         "7/28/14",
         "No",
         "C50.9",
         "8500/3",
         "C50.9",
         "Yes",
         "No",
         "M0",
         "N0 (I+)",
         "T1C",
         "Tumor Free",
         "Yes",
         "No",
         "Black or African American",
         "No",
         null,
         "Yes",
         "0:LIVING",
         "47.60495775",
         "0:ALIVE OR DEAD TUMOR FREE",
         "47.60495775",
         "unknown",
         null,
         "0:CENSORED",
         "47.60495775",
         "AFR",
         "2334_label_0.png"
        ],
        [
         "TCGA-4H-AAAK-01",
         "ILC",
         "Breast Cancer",
         "Breast Invasive Lobular Carcinoma",
         "Infiltrating Lobular Carcinoma",
         "unknown",
         "Yes",
         "No",
         "4H",
         "Breast",
         "7.0",
         "Primary",
         "0.3411",
         "0.01",
         "Matched",
         "0.7",
         "Proteogenex, Inc.",
         "36.0",
         "BRCA_LumA",
         "BRCA",
         "6623FC5E-00BE-4476-967A-CBD55F676EA6",
         "50.0",
         "Female",
         "STAGE IIIA",
         "7TH",
         "348.0",
         "-18371.0",
         "0.0",
         "Not Hispanic Or Latino",
         "11/13/14",
         "No",
         "C50.9",
         "8520/3",
         "C50.9",
         "Yes",
         "No",
         "M0",
         "N2A",
         "T2",
         "Tumor Free",
         "Yes",
         "No",
         "White",
         "No",
         null,
         "Yes",
         "0:LIVING",
         "11.44097051",
         "0:ALIVE OR DEAD TUMOR FREE",
         "11.44097051",
         "0:DiseaseFree",
         "11.44097051",
         "0:CENSORED",
         "11.44097051",
         "EUR",
         "2335_label_0.png"
        ]
       ],
       "shape": {
        "columns": 55,
        "rows": 5
       }
      },
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ONCOTREE_CODE</th>\n",
       "      <th>CANCER_TYPE</th>\n",
       "      <th>CANCER_TYPE_DETAILED</th>\n",
       "      <th>TUMOR_TYPE</th>\n",
       "      <th>GRADE</th>\n",
       "      <th>TISSUE_PROSPECTIVE_COLLECTION_INDICATOR</th>\n",
       "      <th>TISSUE_RETROSPECTIVE_COLLECTION_INDICATOR</th>\n",
       "      <th>TISSUE_SOURCE_SITE_CODE</th>\n",
       "      <th>TUMOR_TISSUE_SITE</th>\n",
       "      <th>ANEUPLOIDY_SCORE</th>\n",
       "      <th>...</th>\n",
       "      <th>OS_STATUS</th>\n",
       "      <th>OS_MONTHS</th>\n",
       "      <th>DSS_STATUS</th>\n",
       "      <th>DSS_MONTHS</th>\n",
       "      <th>DFS_STATUS</th>\n",
       "      <th>DFS_MONTHS</th>\n",
       "      <th>PFS_STATUS</th>\n",
       "      <th>PFS_MONTHS</th>\n",
       "      <th>GENETIC_ANCESTRY_LABEL</th>\n",
       "      <th>img_paths</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMPLE_ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>TCGA-3C-AAAU-01</th>\n",
       "      <td>ILC</td>\n",
       "      <td>Breast Cancer</td>\n",
       "      <td>Breast Invasive Lobular Carcinoma</td>\n",
       "      <td>Infiltrating Lobular Carcinoma</td>\n",
       "      <td>unknown</td>\n",
       "      <td>No</td>\n",
       "      <td>Yes</td>\n",
       "      <td>3C</td>\n",
       "      <td>Breast</td>\n",
       "      <td>19.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0:LIVING</td>\n",
       "      <td>133.050597</td>\n",
       "      <td>0:ALIVE OR DEAD TUMOR FREE</td>\n",
       "      <td>133.050597</td>\n",
       "      <td>1:Recurred/Progressed</td>\n",
       "      <td>59.440444</td>\n",
       "      <td>1:PROGRESSION</td>\n",
       "      <td>59.440444</td>\n",
       "      <td>EUR</td>\n",
       "      <td>2331_label_0.png</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-3C-AALI-01</th>\n",
       "      <td>IDC</td>\n",
       "      <td>Breast Cancer</td>\n",
       "      <td>Breast Invasive Ductal Carcinoma</td>\n",
       "      <td>Infiltrating Ductal Carcinoma</td>\n",
       "      <td>unknown</td>\n",
       "      <td>No</td>\n",
       "      <td>Yes</td>\n",
       "      <td>3C</td>\n",
       "      <td>Breast</td>\n",
       "      <td>22.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0:LIVING</td>\n",
       "      <td>131.669790</td>\n",
       "      <td>0:ALIVE OR DEAD TUMOR FREE</td>\n",
       "      <td>131.669790</td>\n",
       "      <td>0:DiseaseFree</td>\n",
       "      <td>131.669790</td>\n",
       "      <td>0:CENSORED</td>\n",
       "      <td>131.669790</td>\n",
       "      <td>AFR</td>\n",
       "      <td>2332_label_0.png</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-3C-AALJ-01</th>\n",
       "      <td>IDC</td>\n",
       "      <td>Breast Cancer</td>\n",
       "      <td>Breast Invasive Ductal Carcinoma</td>\n",
       "      <td>Infiltrating Ductal Carcinoma</td>\n",
       "      <td>unknown</td>\n",
       "      <td>No</td>\n",
       "      <td>Yes</td>\n",
       "      <td>3C</td>\n",
       "      <td>Breast</td>\n",
       "      <td>13.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0:LIVING</td>\n",
       "      <td>48.459743</td>\n",
       "      <td>0:ALIVE OR DEAD TUMOR FREE</td>\n",
       "      <td>48.459743</td>\n",
       "      <td>0:DiseaseFree</td>\n",
       "      <td>48.459743</td>\n",
       "      <td>0:CENSORED</td>\n",
       "      <td>48.459743</td>\n",
       "      <td>AFR_ADMIX</td>\n",
       "      <td>2333_label_0.png</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-3C-AALK-01</th>\n",
       "      <td>IDC</td>\n",
       "      <td>Breast Cancer</td>\n",
       "      <td>Breast Invasive Ductal Carcinoma</td>\n",
       "      <td>Infiltrating Ductal Carcinoma</td>\n",
       "      <td>unknown</td>\n",
       "      <td>No</td>\n",
       "      <td>Yes</td>\n",
       "      <td>3C</td>\n",
       "      <td>Breast</td>\n",
       "      <td>4.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0:LIVING</td>\n",
       "      <td>47.604958</td>\n",
       "      <td>0:ALIVE OR DEAD TUMOR FREE</td>\n",
       "      <td>47.604958</td>\n",
       "      <td>unknown</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0:CENSORED</td>\n",
       "      <td>47.604958</td>\n",
       "      <td>AFR</td>\n",
       "      <td>2334_label_0.png</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-4H-AAAK-01</th>\n",
       "      <td>ILC</td>\n",
       "      <td>Breast Cancer</td>\n",
       "      <td>Breast Invasive Lobular Carcinoma</td>\n",
       "      <td>Infiltrating Lobular Carcinoma</td>\n",
       "      <td>unknown</td>\n",
       "      <td>Yes</td>\n",
       "      <td>No</td>\n",
       "      <td>4H</td>\n",
       "      <td>Breast</td>\n",
       "      <td>7.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0:LIVING</td>\n",
       "      <td>11.440971</td>\n",
       "      <td>0:ALIVE OR DEAD TUMOR FREE</td>\n",
       "      <td>11.440971</td>\n",
       "      <td>0:DiseaseFree</td>\n",
       "      <td>11.440971</td>\n",
       "      <td>0:CENSORED</td>\n",
       "      <td>11.440971</td>\n",
       "      <td>EUR</td>\n",
       "      <td>2335_label_0.png</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 55 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                ONCOTREE_CODE    CANCER_TYPE  \\\n",
       "SAMPLE_ID                                      \n",
       "TCGA-3C-AAAU-01           ILC  Breast Cancer   \n",
       "TCGA-3C-AALI-01           IDC  Breast Cancer   \n",
       "TCGA-3C-AALJ-01           IDC  Breast Cancer   \n",
       "TCGA-3C-AALK-01           IDC  Breast Cancer   \n",
       "TCGA-4H-AAAK-01           ILC  Breast Cancer   \n",
       "\n",
       "                              CANCER_TYPE_DETAILED  \\\n",
       "SAMPLE_ID                                            \n",
       "TCGA-3C-AAAU-01  Breast Invasive Lobular Carcinoma   \n",
       "TCGA-3C-AALI-01   Breast Invasive Ductal Carcinoma   \n",
       "TCGA-3C-AALJ-01   Breast Invasive Ductal Carcinoma   \n",
       "TCGA-3C-AALK-01   Breast Invasive Ductal Carcinoma   \n",
       "TCGA-4H-AAAK-01  Breast Invasive Lobular Carcinoma   \n",
       "\n",
       "                                     TUMOR_TYPE    GRADE  \\\n",
       "SAMPLE_ID                                                  \n",
       "TCGA-3C-AAAU-01  Infiltrating Lobular Carcinoma  unknown   \n",
       "TCGA-3C-AALI-01   Infiltrating Ductal Carcinoma  unknown   \n",
       "TCGA-3C-AALJ-01   Infiltrating Ductal Carcinoma  unknown   \n",
       "TCGA-3C-AALK-01   Infiltrating Ductal Carcinoma  unknown   \n",
       "TCGA-4H-AAAK-01  Infiltrating Lobular Carcinoma  unknown   \n",
       "\n",
       "                TISSUE_PROSPECTIVE_COLLECTION_INDICATOR  \\\n",
       "SAMPLE_ID                                                 \n",
       "TCGA-3C-AAAU-01                                      No   \n",
       "TCGA-3C-AALI-01                                      No   \n",
       "TCGA-3C-AALJ-01                                      No   \n",
       "TCGA-3C-AALK-01                                      No   \n",
       "TCGA-4H-AAAK-01                                     Yes   \n",
       "\n",
       "                TISSUE_RETROSPECTIVE_COLLECTION_INDICATOR  \\\n",
       "SAMPLE_ID                                                   \n",
       "TCGA-3C-AAAU-01                                       Yes   \n",
       "TCGA-3C-AALI-01                                       Yes   \n",
       "TCGA-3C-AALJ-01                                       Yes   \n",
       "TCGA-3C-AALK-01                                       Yes   \n",
       "TCGA-4H-AAAK-01                                        No   \n",
       "\n",
       "                TISSUE_SOURCE_SITE_CODE TUMOR_TISSUE_SITE  ANEUPLOIDY_SCORE  \\\n",
       "SAMPLE_ID                                                                     \n",
       "TCGA-3C-AAAU-01                      3C            Breast              19.0   \n",
       "TCGA-3C-AALI-01                      3C            Breast              22.0   \n",
       "TCGA-3C-AALJ-01                      3C            Breast              13.0   \n",
       "TCGA-3C-AALK-01                      3C            Breast               4.0   \n",
       "TCGA-4H-AAAK-01                      4H            Breast               7.0   \n",
       "\n",
       "                 ... OS_STATUS   OS_MONTHS                  DSS_STATUS  \\\n",
       "SAMPLE_ID        ...                                                     \n",
       "TCGA-3C-AAAU-01  ...  0:LIVING  133.050597  0:ALIVE OR DEAD TUMOR FREE   \n",
       "TCGA-3C-AALI-01  ...  0:LIVING  131.669790  0:ALIVE OR DEAD TUMOR FREE   \n",
       "TCGA-3C-AALJ-01  ...  0:LIVING   48.459743  0:ALIVE OR DEAD TUMOR FREE   \n",
       "TCGA-3C-AALK-01  ...  0:LIVING   47.604958  0:ALIVE OR DEAD TUMOR FREE   \n",
       "TCGA-4H-AAAK-01  ...  0:LIVING   11.440971  0:ALIVE OR DEAD TUMOR FREE   \n",
       "\n",
       "                 DSS_MONTHS             DFS_STATUS  DFS_MONTHS     PFS_STATUS  \\\n",
       "SAMPLE_ID                                                                       \n",
       "TCGA-3C-AAAU-01  133.050597  1:Recurred/Progressed   59.440444  1:PROGRESSION   \n",
       "TCGA-3C-AALI-01  131.669790          0:DiseaseFree  131.669790     0:CENSORED   \n",
       "TCGA-3C-AALJ-01   48.459743          0:DiseaseFree   48.459743     0:CENSORED   \n",
       "TCGA-3C-AALK-01   47.604958                unknown         NaN     0:CENSORED   \n",
       "TCGA-4H-AAAK-01   11.440971          0:DiseaseFree   11.440971     0:CENSORED   \n",
       "\n",
       "                 PFS_MONTHS GENETIC_ANCESTRY_LABEL         img_paths  \n",
       "SAMPLE_ID                                                             \n",
       "TCGA-3C-AAAU-01   59.440444                    EUR  2331_label_0.png  \n",
       "TCGA-3C-AALI-01  131.669790                    AFR  2332_label_0.png  \n",
       "TCGA-3C-AALJ-01   48.459743              AFR_ADMIX  2333_label_0.png  \n",
       "TCGA-3C-AALK-01   47.604958                    AFR  2334_label_0.png  \n",
       "TCGA-4H-AAAK-01   11.440971                    EUR  2335_label_0.png  \n",
       "\n",
       "[5 rows x 55 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "import autoencodix as acx\n",
    "\n",
    "from autoencodix.configs.xmodalix_config import XModalixConfig\n",
    "from autoencodix.configs.default_config import DataConfig, DataInfo, DataCase\n",
    "\n",
    "clin_file = os.path.join(\"data/XModalix-Tut-data/combined_clin_formatted.parquet\")\n",
    "clin_df = pd.read_parquet(clin_file)\n",
    "print(\"First five image paths:\")\n",
    "print(clin_df[\"img_paths\"].to_list()[0:5])\n",
    "print(\"\\n\")\n",
    "clin_df.head()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5fed079a",
   "metadata": {},
   "source": [
    "Another way is to provide a image-specific annotation file in with the `extra_anno_file` parameter of the DataInfo object in the DataConfig.\n",
    "Here you need to map the sample_id (index) to the image path and can add addtional metadata columns like we did in our `imganno_file`.   \n",
    "> We recommend the first option of a global annotation file.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "b11dfba0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.microsoft.datawrangler.viewer.v0+json": {
       "columns": [
        {
         "name": "sample_ids",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "img_paths",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "extra_class_labels",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "CANCER_TYPE_ACRONYM",
         "rawType": "object",
         "type": "string"
        }
       ],
       "ref": "633f8ac3-e7ac-46a9-beb9-c36fe89c6347",
       "rows": [
        [
         "TCGA-05-4244-01",
         "0_label_1.png",
         "Non-Small Cell Lung Cancer",
         "LUAD"
        ],
        [
         "TCGA-05-4249-01",
         "1_label_1.png",
         "Non-Small Cell Lung Cancer",
         "LUAD"
        ],
        [
         "TCGA-05-4250-01",
         "2_label_1.png",
         "Non-Small Cell Lung Cancer",
         "LUAD"
        ],
        [
         "TCGA-05-4382-01",
         "3_label_1.png",
         "Non-Small Cell Lung Cancer",
         "LUAD"
        ],
        [
         "TCGA-05-4384-01",
         "4_label_1.png",
         "Non-Small Cell Lung Cancer",
         "LUAD"
        ]
       ],
       "shape": {
        "columns": 3,
        "rows": 5
       }
      },
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>img_paths</th>\n",
       "      <th>extra_class_labels</th>\n",
       "      <th>CANCER_TYPE_ACRONYM</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample_ids</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>TCGA-05-4244-01</th>\n",
       "      <td>0_label_1.png</td>\n",
       "      <td>Non-Small Cell Lung Cancer</td>\n",
       "      <td>LUAD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-05-4249-01</th>\n",
       "      <td>1_label_1.png</td>\n",
       "      <td>Non-Small Cell Lung Cancer</td>\n",
       "      <td>LUAD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-05-4250-01</th>\n",
       "      <td>2_label_1.png</td>\n",
       "      <td>Non-Small Cell Lung Cancer</td>\n",
       "      <td>LUAD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-05-4382-01</th>\n",
       "      <td>3_label_1.png</td>\n",
       "      <td>Non-Small Cell Lung Cancer</td>\n",
       "      <td>LUAD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-05-4384-01</th>\n",
       "      <td>4_label_1.png</td>\n",
       "      <td>Non-Small Cell Lung Cancer</td>\n",
       "      <td>LUAD</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                     img_paths          extra_class_labels CANCER_TYPE_ACRONYM\n",
       "sample_ids                                                                    \n",
       "TCGA-05-4244-01  0_label_1.png  Non-Small Cell Lung Cancer                LUAD\n",
       "TCGA-05-4249-01  1_label_1.png  Non-Small Cell Lung Cancer                LUAD\n",
       "TCGA-05-4250-01  2_label_1.png  Non-Small Cell Lung Cancer                LUAD\n",
       "TCGA-05-4382-01  3_label_1.png  Non-Small Cell Lung Cancer                LUAD\n",
       "TCGA-05-4384-01  4_label_1.png  Non-Small Cell Lung Cancer                LUAD"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "imganno_file = os.path.join(\"data/XModalix-Tut-data/tcga_mappings.txt\")\n",
    "img_anno_df = pd.read_csv(imganno_file, sep=\"\\t\", index_col=0)\n",
    "img_anno_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a96e186f",
   "metadata": {},
   "source": [
    "## 1) Initialize and Run XModalix\n",
    "\n",
    "In this section, we demonstrate how to set up the configuration to account for the data requirements and highlight the custom XModalix parameters.  \n",
    "\n",
    "For a more in-depth guide on the configuration parameters and input data, see:  \n",
    "\n",
    "[1] `Tutorials/DeepDives/ConfigTutorial.ipynb`  \n",
    "[2] `Tutorials/DeepDives/InputDataTutorials.ipynb`\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "593e79b0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "reading parquet: data/XModalix-Tut-data/combined_rnaseq_formatted.parquet\n",
      "reading parquet: data/XModalix-Tut-data/combined_clin_formatted.parquet\n",
      "Given image size is possible, rescaling images to: 64x64\n",
      "Successfully loaded 3230 images for img\n",
      "anno key: rna\n",
      "anno key: img\n",
      "Converting 2261 images to torch.float32 tensors...\n",
      "Converting 646 images to torch.float32 tensors...\n",
      "Converting 323 images to torch.float32 tensors...\n",
      "key: train, type: <class 'dict'>\n",
      "key: valid, type: <class 'dict'>\n",
      "key: test, type: <class 'dict'>\n",
      "Check if we need to pretrain: multi_bulk.rna\n",
      "pretrain epochs : 20\n",
      "Starting Pretraining for: multi_bulk.rna with <class 'autoencodix.trainers._general_trainer.GeneralTrainer'>\n",
      "Epoch 1 - Train Loss: 4.4723\n",
      "Sub-losses: recon_loss: 4.4723, var_loss: 0.0000, anneal_factor: 0.0000, effective_beta_factor: 0.0000\n",
      "Epoch 1 - Valid Loss: 1.4290\n",
      "Sub-losses: recon_loss: 1.4290, var_loss: 0.0000, anneal_factor: 0.0000, effective_beta_factor: 0.0000\n",
      "Epoch 2 - Train Loss: 1.6831\n",
      "Sub-losses: recon_loss: 1.6831, var_loss: 0.0000, anneal_factor: 0.0001, effective_beta_factor: 0.0000\n",
      "Epoch 2 - Valid Loss: 0.6086\n",
      "Sub-losses: recon_loss: 0.6085, var_loss: 0.0000, anneal_factor: 0.0001, effective_beta_factor: 0.0000\n",
      "Epoch 3 - Train Loss: 0.8171\n",
      "Sub-losses: recon_loss: 0.8171, var_loss: 0.0000, anneal_factor: 0.0002, effective_beta_factor: 0.0000\n",
      "Epoch 3 - Valid Loss: 0.3717\n",
      "Sub-losses: recon_loss: 0.3717, var_loss: 0.0000, anneal_factor: 0.0002, effective_beta_factor: 0.0000\n",
      "Epoch 4 - Train Loss: 0.5464\n",
      "Sub-losses: recon_loss: 0.5463, var_loss: 0.0000, anneal_factor: 0.0003, effective_beta_factor: 0.0000\n",
      "Epoch 4 - Valid Loss: 0.2290\n",
      "Sub-losses: recon_loss: 0.2290, var_loss: 0.0000, anneal_factor: 0.0003, effective_beta_factor: 0.0000\n",
      "Epoch 5 - Train Loss: 0.3970\n",
      "Sub-losses: recon_loss: 0.3969, var_loss: 0.0000, anneal_factor: 0.0007, effective_beta_factor: 0.0000\n",
      "Epoch 5 - Valid Loss: 0.2216\n",
      "Sub-losses: recon_loss: 0.2216, var_loss: 0.0000, anneal_factor: 0.0007, effective_beta_factor: 0.0000\n",
      "Epoch 6 - Train Loss: 0.3243\n",
      "Sub-losses: recon_loss: 0.3242, var_loss: 0.0000, anneal_factor: 0.0013, effective_beta_factor: 0.0000\n",
      "Epoch 6 - Valid Loss: 0.1839\n",
      "Sub-losses: recon_loss: 0.1839, var_loss: 0.0000, anneal_factor: 0.0013, effective_beta_factor: 0.0000\n",
      "Epoch 7 - Train Loss: 0.2845\n",
      "Sub-losses: recon_loss: 0.2844, var_loss: 0.0001, anneal_factor: 0.0025, effective_beta_factor: 0.0000\n",
      "Epoch 7 - Valid Loss: 0.1768\n",
      "Sub-losses: recon_loss: 0.1767, var_loss: 0.0001, anneal_factor: 0.0025, effective_beta_factor: 0.0000\n",
      "Epoch 8 - Train Loss: 0.2411\n",
      "Sub-losses: recon_loss: 0.2409, var_loss: 0.0002, anneal_factor: 0.0048, effective_beta_factor: 0.0000\n",
      "Epoch 8 - Valid Loss: 0.1506\n",
      "Sub-losses: recon_loss: 0.1505, var_loss: 0.0001, anneal_factor: 0.0048, effective_beta_factor: 0.0000\n",
      "Epoch 9 - Train Loss: 0.2204\n",
      "Sub-losses: recon_loss: 0.2200, var_loss: 0.0004, anneal_factor: 0.0093, effective_beta_factor: 0.0001\n",
      "Epoch 9 - Valid Loss: 0.1490\n",
      "Sub-losses: recon_loss: 0.1488, var_loss: 0.0002, anneal_factor: 0.0093, effective_beta_factor: 0.0001\n",
      "Epoch 10 - Train Loss: 0.1961\n",
      "Sub-losses: recon_loss: 0.1954, var_loss: 0.0007, anneal_factor: 0.0180, effective_beta_factor: 0.0002\n",
      "Epoch 10 - Valid Loss: 0.1487\n",
      "Sub-losses: recon_loss: 0.1483, var_loss: 0.0004, anneal_factor: 0.0180, effective_beta_factor: 0.0002\n",
      "Epoch 11 - Train Loss: 0.1853\n",
      "Sub-losses: recon_loss: 0.1839, var_loss: 0.0014, anneal_factor: 0.0344, effective_beta_factor: 0.0003\n",
      "Epoch 11 - Valid Loss: 0.1409\n",
      "Sub-losses: recon_loss: 0.1401, var_loss: 0.0008, anneal_factor: 0.0344, effective_beta_factor: 0.0003\n",
      "Epoch 12 - Train Loss: 0.1767\n",
      "Sub-losses: recon_loss: 0.1741, var_loss: 0.0026, anneal_factor: 0.0650, effective_beta_factor: 0.0006\n",
      "Epoch 12 - Valid Loss: 0.1414\n",
      "Sub-losses: recon_loss: 0.1404, var_loss: 0.0010, anneal_factor: 0.0650, effective_beta_factor: 0.0006\n",
      "Epoch 13 - Train Loss: 0.1699\n",
      "Sub-losses: recon_loss: 0.1649, var_loss: 0.0050, anneal_factor: 0.1192, effective_beta_factor: 0.0012\n",
      "Epoch 13 - Valid Loss: 0.1390\n",
      "Sub-losses: recon_loss: 0.1370, var_loss: 0.0020, anneal_factor: 0.1192, effective_beta_factor: 0.0012\n",
      "Epoch 14 - Train Loss: 0.1678\n",
      "Sub-losses: recon_loss: 0.1593, var_loss: 0.0086, anneal_factor: 0.2086, effective_beta_factor: 0.0021\n",
      "Epoch 14 - Valid Loss: 0.1404\n",
      "Sub-losses: recon_loss: 0.1351, var_loss: 0.0054, anneal_factor: 0.2086, effective_beta_factor: 0.0021\n",
      "Epoch 15 - Train Loss: 0.1643\n",
      "Sub-losses: recon_loss: 0.1513, var_loss: 0.0130, anneal_factor: 0.3392, effective_beta_factor: 0.0034\n",
      "Epoch 15 - Valid Loss: 0.1414\n",
      "Sub-losses: recon_loss: 0.1336, var_loss: 0.0078, anneal_factor: 0.3392, effective_beta_factor: 0.0034\n",
      "Epoch 16 - Train Loss: 0.1663\n",
      "Sub-losses: recon_loss: 0.1477, var_loss: 0.0186, anneal_factor: 0.5000, effective_beta_factor: 0.0050\n",
      "Epoch 16 - Valid Loss: 0.1469\n",
      "Sub-losses: recon_loss: 0.1328, var_loss: 0.0141, anneal_factor: 0.5000, effective_beta_factor: 0.0050\n",
      "Epoch 17 - Train Loss: 0.1625\n",
      "Sub-losses: recon_loss: 0.1400, var_loss: 0.0226, anneal_factor: 0.6608, effective_beta_factor: 0.0066\n",
      "Epoch 17 - Valid Loss: 0.1419\n",
      "Sub-losses: recon_loss: 0.1332, var_loss: 0.0087, anneal_factor: 0.6608, effective_beta_factor: 0.0066\n",
      "Epoch 18 - Train Loss: 0.1617\n",
      "Sub-losses: recon_loss: 0.1384, var_loss: 0.0233, anneal_factor: 0.7914, effective_beta_factor: 0.0079\n",
      "Epoch 18 - Valid Loss: 0.1400\n",
      "Sub-losses: recon_loss: 0.1306, var_loss: 0.0094, anneal_factor: 0.7914, effective_beta_factor: 0.0079\n",
      "Epoch 19 - Train Loss: 0.1575\n",
      "Sub-losses: recon_loss: 0.1358, var_loss: 0.0217, anneal_factor: 0.8808, effective_beta_factor: 0.0088\n",
      "Epoch 19 - Valid Loss: 0.1443\n",
      "Sub-losses: recon_loss: 0.1303, var_loss: 0.0140, anneal_factor: 0.8808, effective_beta_factor: 0.0088\n",
      "Epoch 20 - Train Loss: 0.1533\n",
      "Sub-losses: recon_loss: 0.1330, var_loss: 0.0203, anneal_factor: 0.9350, effective_beta_factor: 0.0094\n",
      "Epoch 20 - Valid Loss: 0.1396\n",
      "Sub-losses: recon_loss: 0.1289, var_loss: 0.0106, anneal_factor: 0.9350, effective_beta_factor: 0.0094\n",
      "Check if we need to pretrain: img.img\n",
      "pretrain epochs : 20\n",
      "Starting Pretraining for: img.img with <class 'autoencodix.trainers._general_trainer.GeneralTrainer'>\n",
      "Epoch 1 - Train Loss: 508.2926\n",
      "Sub-losses: recon_loss: 508.1462, var_loss: 0.1464, anneal_factor: 0.0000, effective_beta_factor: 0.0000\n",
      "Epoch 1 - Valid Loss: 403.6758\n",
      "Sub-losses: recon_loss: 403.6754, var_loss: 0.0004, anneal_factor: 0.0000, effective_beta_factor: 0.0000\n",
      "Epoch 2 - Train Loss: 223.1287\n",
      "Sub-losses: recon_loss: 223.1283, var_loss: 0.0004, anneal_factor: 0.0001, effective_beta_factor: 0.0000\n",
      "Epoch 2 - Valid Loss: 177.6630\n",
      "Sub-losses: recon_loss: 177.6627, var_loss: 0.0003, anneal_factor: 0.0001, effective_beta_factor: 0.0000\n",
      "Epoch 3 - Train Loss: 168.6337\n",
      "Sub-losses: recon_loss: 168.6332, var_loss: 0.0005, anneal_factor: 0.0002, effective_beta_factor: 0.0000\n",
      "Epoch 3 - Valid Loss: 143.9080\n",
      "Sub-losses: recon_loss: 143.9074, var_loss: 0.0006, anneal_factor: 0.0002, effective_beta_factor: 0.0000\n",
      "Epoch 4 - Train Loss: 140.5889\n",
      "Sub-losses: recon_loss: 140.5879, var_loss: 0.0011, anneal_factor: 0.0003, effective_beta_factor: 0.0000\n",
      "Epoch 4 - Valid Loss: 124.6864\n",
      "Sub-losses: recon_loss: 124.6852, var_loss: 0.0012, anneal_factor: 0.0003, effective_beta_factor: 0.0000\n",
      "Epoch 5 - Train Loss: 124.6277\n",
      "Sub-losses: recon_loss: 124.6249, var_loss: 0.0027, anneal_factor: 0.0007, effective_beta_factor: 0.0000\n",
      "Epoch 5 - Valid Loss: 111.1609\n",
      "Sub-losses: recon_loss: 111.1577, var_loss: 0.0031, anneal_factor: 0.0007, effective_beta_factor: 0.0000\n",
      "Epoch 6 - Train Loss: 114.2849\n",
      "Sub-losses: recon_loss: 114.2783, var_loss: 0.0066, anneal_factor: 0.0013, effective_beta_factor: 0.0000\n",
      "Epoch 6 - Valid Loss: 104.9094\n",
      "Sub-losses: recon_loss: 104.9021, var_loss: 0.0073, anneal_factor: 0.0013, effective_beta_factor: 0.0000\n",
      "Epoch 7 - Train Loss: 108.6431\n",
      "Sub-losses: recon_loss: 108.6276, var_loss: 0.0155, anneal_factor: 0.0025, effective_beta_factor: 0.0000\n",
      "Epoch 7 - Valid Loss: 100.8577\n",
      "Sub-losses: recon_loss: 100.8407, var_loss: 0.0171, anneal_factor: 0.0025, effective_beta_factor: 0.0000\n",
      "Epoch 8 - Train Loss: 103.4417\n",
      "Sub-losses: recon_loss: 103.4067, var_loss: 0.0350, anneal_factor: 0.0048, effective_beta_factor: 0.0000\n",
      "Epoch 8 - Valid Loss: 99.7484\n",
      "Sub-losses: recon_loss: 99.7084, var_loss: 0.0399, anneal_factor: 0.0048, effective_beta_factor: 0.0000\n",
      "Epoch 9 - Train Loss: 98.6755\n",
      "Sub-losses: recon_loss: 98.5994, var_loss: 0.0761, anneal_factor: 0.0093, effective_beta_factor: 0.0001\n",
      "Epoch 9 - Valid Loss: 96.1943\n",
      "Sub-losses: recon_loss: 96.1155, var_loss: 0.0788, anneal_factor: 0.0093, effective_beta_factor: 0.0001\n",
      "Epoch 10 - Train Loss: 95.6776\n",
      "Sub-losses: recon_loss: 95.5159, var_loss: 0.1617, anneal_factor: 0.0180, effective_beta_factor: 0.0002\n",
      "Epoch 10 - Valid Loss: 92.3834\n",
      "Sub-losses: recon_loss: 92.2178, var_loss: 0.1656, anneal_factor: 0.0180, effective_beta_factor: 0.0002\n",
      "Epoch 11 - Train Loss: 91.7394\n",
      "Sub-losses: recon_loss: 91.3992, var_loss: 0.3403, anneal_factor: 0.0344, effective_beta_factor: 0.0003\n",
      "Epoch 11 - Valid Loss: 91.0587\n",
      "Sub-losses: recon_loss: 90.6697, var_loss: 0.3891, anneal_factor: 0.0344, effective_beta_factor: 0.0003\n",
      "Epoch 12 - Train Loss: 88.0785\n",
      "Sub-losses: recon_loss: 87.3867, var_loss: 0.6918, anneal_factor: 0.0650, effective_beta_factor: 0.0006\n",
      "Epoch 12 - Valid Loss: 86.8266\n",
      "Sub-losses: recon_loss: 86.0882, var_loss: 0.7384, anneal_factor: 0.0650, effective_beta_factor: 0.0006\n",
      "Epoch 13 - Train Loss: 83.3224\n",
      "Sub-losses: recon_loss: 82.0367, var_loss: 1.2857, anneal_factor: 0.1192, effective_beta_factor: 0.0012\n",
      "Epoch 13 - Valid Loss: 83.8232\n",
      "Sub-losses: recon_loss: 82.5140, var_loss: 1.3092, anneal_factor: 0.1192, effective_beta_factor: 0.0012\n",
      "Epoch 14 - Train Loss: 82.3940\n",
      "Sub-losses: recon_loss: 80.0626, var_loss: 2.3314, anneal_factor: 0.2086, effective_beta_factor: 0.0021\n",
      "Epoch 14 - Valid Loss: 83.5220\n",
      "Sub-losses: recon_loss: 81.2827, var_loss: 2.2393, anneal_factor: 0.2086, effective_beta_factor: 0.0021\n",
      "Epoch 15 - Train Loss: 80.6536\n",
      "Sub-losses: recon_loss: 76.9564, var_loss: 3.6972, anneal_factor: 0.3392, effective_beta_factor: 0.0034\n",
      "Epoch 15 - Valid Loss: 81.1522\n",
      "Sub-losses: recon_loss: 77.4612, var_loss: 3.6910, anneal_factor: 0.3392, effective_beta_factor: 0.0034\n",
      "Epoch 16 - Train Loss: 80.9892\n",
      "Sub-losses: recon_loss: 75.9855, var_loss: 5.0037, anneal_factor: 0.5000, effective_beta_factor: 0.0050\n",
      "Epoch 16 - Valid Loss: 80.7657\n",
      "Sub-losses: recon_loss: 76.0288, var_loss: 4.7369, anneal_factor: 0.5000, effective_beta_factor: 0.0050\n",
      "Epoch 17 - Train Loss: 80.0775\n",
      "Sub-losses: recon_loss: 74.2579, var_loss: 5.8196, anneal_factor: 0.6608, effective_beta_factor: 0.0066\n",
      "Epoch 17 - Valid Loss: 83.9216\n",
      "Sub-losses: recon_loss: 78.3710, var_loss: 5.5506, anneal_factor: 0.6608, effective_beta_factor: 0.0066\n",
      "Epoch 18 - Train Loss: 81.1761\n",
      "Sub-losses: recon_loss: 75.0749, var_loss: 6.1012, anneal_factor: 0.7914, effective_beta_factor: 0.0079\n",
      "Epoch 18 - Valid Loss: 83.1751\n",
      "Sub-losses: recon_loss: 77.7133, var_loss: 5.4617, anneal_factor: 0.7914, effective_beta_factor: 0.0079\n",
      "Epoch 19 - Train Loss: 79.4787\n",
      "Sub-losses: recon_loss: 73.2670, var_loss: 6.2117, anneal_factor: 0.8808, effective_beta_factor: 0.0088\n",
      "Epoch 19 - Valid Loss: 85.8535\n",
      "Sub-losses: recon_loss: 79.5687, var_loss: 6.2847, anneal_factor: 0.8808, effective_beta_factor: 0.0088\n",
      "Epoch 20 - Train Loss: 80.0211\n",
      "Sub-losses: recon_loss: 73.7558, var_loss: 6.2653, anneal_factor: 0.9350, effective_beta_factor: 0.0094\n",
      "Epoch 20 - Valid Loss: 83.1510\n",
      "Sub-losses: recon_loss: 77.3115, var_loss: 5.8394, anneal_factor: 0.9350, effective_beta_factor: 0.0094\n",
      "--- Epoch 1/30 ---\n",
      "split: train, n_samples: 11530\n",
      "Epoch 1/30 - Train Loss: 50.0151\n",
      "Sub-losses - total_loss: 50.0151, adver_loss: 6.1649, aggregated_sub_losses: 28.9324, paired_loss: 10.7716, class_loss: 4.1462, multi_bulk.rna.recon_loss: 0.0765, multi_bulk.rna.var_loss: 0.0000, multi_bulk.rna.loss: 0.0765, img.img.recon_loss: 28.8558, img.img.var_loss: 0.0001, img.img.loss: 28.8559, clf_loss: 0.1618\n",
      "split: valid, n_samples: 1681\n",
      "Epoch 1/30 - Valid Loss: 92099.3484\n",
      "Sub-losses - total_loss: 54.7884, adver_loss: 8.9235, aggregated_sub_losses: 30.9227, paired_loss: 10.8233, class_loss: 4.1189, multi_bulk.rna.recon_loss: 0.0793, multi_bulk.rna.var_loss: 0.0000, multi_bulk.rna.loss: 0.0793, img.img.recon_loss: 30.8433, img.img.var_loss: 0.0001, img.img.loss: 30.8434, clf_loss: 0.0197\n",
      "--- Epoch 2/30 ---\n",
      "split: train, n_samples: 11530\n",
      "Epoch 2/30 - Train Loss: 54.7916\n",
      "Sub-losses - total_loss: 54.7916, adver_loss: 10.9187, aggregated_sub_losses: 28.9446, paired_loss: 10.7800, class_loss: 4.1483, multi_bulk.rna.recon_loss: 0.0760, multi_bulk.rna.var_loss: 0.0000, multi_bulk.rna.loss: 0.0760, img.img.recon_loss: 28.8683, img.img.var_loss: 0.0002, img.img.loss: 28.8686, clf_loss: 0.0182\n",
      "split: valid, n_samples: 1681\n",
      "Epoch 2/30 - Valid Loss: 97072.1753\n",
      "Sub-losses - total_loss: 57.7467, adver_loss: 12.8068, aggregated_sub_losses: 30.2722, paired_loss: 10.5969, class_loss: 4.0707, multi_bulk.rna.recon_loss: 0.0772, multi_bulk.rna.var_loss: 0.0000, multi_bulk.rna.loss: 0.0772, img.img.recon_loss: 30.1948, img.img.var_loss: 0.0002, img.img.loss: 30.1950, clf_loss: 0.0075\n",
      "--- Epoch 3/30 ---\n",
      "split: train, n_samples: 11530\n",
      "Epoch 3/30 - Train Loss: 57.4487\n",
      "Sub-losses - total_loss: 57.4487, adver_loss: 13.6338, aggregated_sub_losses: 28.9046, paired_loss: 10.7637, class_loss: 4.1466, multi_bulk.rna.recon_loss: 0.0757, multi_bulk.rna.var_loss: 0.0000, multi_bulk.rna.loss: 0.0757, img.img.recon_loss: 28.8285, img.img.var_loss: 0.0004, img.img.loss: 28.8289, clf_loss: 0.0092\n",
      "split: valid, n_samples: 1681\n",
      "Epoch 3/30 - Valid Loss: 101660.8914\n",
      "Sub-losses - total_loss: 60.4764, adver_loss: 15.4041, aggregated_sub_losses: 30.3523, paired_loss: 10.6483, class_loss: 4.0717, multi_bulk.rna.recon_loss: 0.0765, multi_bulk.rna.var_loss: 0.0000, multi_bulk.rna.loss: 0.0765, img.img.recon_loss: 30.2753, img.img.var_loss: 0.0004, img.img.loss: 30.2757, clf_loss: 0.0041\n",
      "--- Epoch 4/30 ---\n",
      "split: train, n_samples: 11530\n",
      "Epoch 4/30 - Train Loss: 59.0242\n",
      "Sub-losses - total_loss: 59.0242, adver_loss: 15.0198, aggregated_sub_losses: 29.0785, paired_loss: 10.7785, class_loss: 4.1474, multi_bulk.rna.recon_loss: 0.0754, multi_bulk.rna.var_loss: 0.0000, multi_bulk.rna.loss: 0.0754, img.img.recon_loss: 29.0022, img.img.var_loss: 0.0009, img.img.loss: 29.0031, clf_loss: 0.0043\n",
      "split: valid, n_samples: 1681\n",
      "Epoch 4/30 - Valid Loss: 101512.9744\n",
      "Sub-losses - total_loss: 60.3884, adver_loss: 15.0422, aggregated_sub_losses: 30.4422, paired_loss: 10.7708, class_loss: 4.1333, multi_bulk.rna.recon_loss: 0.0794, multi_bulk.rna.var_loss: 0.0000, multi_bulk.rna.loss: 0.0794, img.img.recon_loss: 30.3619, img.img.var_loss: 0.0009, img.img.loss: 30.3627, clf_loss: 0.0022\n",
      "--- Epoch 5/30 ---\n",
      "split: train, n_samples: 11530\n",
      "Epoch 5/30 - Train Loss: 59.5870\n",
      "Sub-losses - total_loss: 59.5870, adver_loss: 15.5964, aggregated_sub_losses: 29.0871, paired_loss: 10.7605, class_loss: 4.1430, multi_bulk.rna.recon_loss: 0.0768, multi_bulk.rna.var_loss: 0.0000, multi_bulk.rna.loss: 0.0768, img.img.recon_loss: 29.0086, img.img.var_loss: 0.0017, img.img.loss: 29.0103, clf_loss: 0.0018\n",
      "split: valid, n_samples: 1681\n",
      "Epoch 5/30 - Valid Loss: 100855.0300\n",
      "Sub-losses - total_loss: 59.9970, adver_loss: 15.2243, aggregated_sub_losses: 30.0539, paired_loss: 10.6458, class_loss: 4.0730, multi_bulk.rna.recon_loss: 0.0763, multi_bulk.rna.var_loss: 0.0000, multi_bulk.rna.loss: 0.0763, img.img.recon_loss: 29.9760, img.img.var_loss: 0.0016, img.img.loss: 29.9776, clf_loss: 0.0003\n",
      "Storing checkpoint for epoch 4...\n",
      "--- Epoch 6/30 ---\n",
      "split: train, n_samples: 11530\n",
      "Epoch 6/30 - Train Loss: 60.8178\n",
      "Sub-losses - total_loss: 60.8178, adver_loss: 16.3087, aggregated_sub_losses: 29.1054, paired_loss: 10.7814, class_loss: 4.6224, multi_bulk.rna.recon_loss: 0.0757, multi_bulk.rna.var_loss: 0.0000, multi_bulk.rna.loss: 0.0757, img.img.recon_loss: 29.0264, img.img.var_loss: 0.0032, img.img.loss: 29.0297, clf_loss: 0.0019\n",
      "split: valid, n_samples: 1681\n",
      "Epoch 6/30 - Valid Loss: 103971.9053\n",
      "Sub-losses - total_loss: 61.8512, adver_loss: 16.0311, aggregated_sub_losses: 30.5640, paired_loss: 10.6882, class_loss: 4.5679, multi_bulk.rna.recon_loss: 0.0785, multi_bulk.rna.var_loss: 0.0000, multi_bulk.rna.loss: 0.0785, img.img.recon_loss: 30.4823, img.img.var_loss: 0.0032, img.img.loss: 30.4855, clf_loss: 0.0033\n",
      "--- Epoch 7/30 ---\n",
      "split: train, n_samples: 11530\n",
      "Epoch 7/30 - Train Loss: 60.9082\n",
      "Sub-losses - total_loss: 60.9082, adver_loss: 16.4708, aggregated_sub_losses: 29.0446, paired_loss: 10.7667, class_loss: 4.6260, multi_bulk.rna.recon_loss: 0.0766, multi_bulk.rna.var_loss: 0.0000, multi_bulk.rna.loss: 0.0767, img.img.recon_loss: 28.9616, img.img.var_loss: 0.0063, img.img.loss: 28.9680, clf_loss: 0.0012\n",
      "split: valid, n_samples: 1681\n",
      "Epoch 7/30 - Valid Loss: 106229.0183\n",
      "Sub-losses - total_loss: 63.1939, adver_loss: 17.6001, aggregated_sub_losses: 30.3572, paired_loss: 10.6620, class_loss: 4.5746, multi_bulk.rna.recon_loss: 0.0773, multi_bulk.rna.var_loss: 0.0000, multi_bulk.rna.loss: 0.0773, img.img.recon_loss: 30.2738, img.img.var_loss: 0.0061, img.img.loss: 30.2799, clf_loss: 0.0001\n",
      "--- Epoch 8/30 ---\n",
      "split: train, n_samples: 11530\n",
      "Epoch 8/30 - Train Loss: 61.9073\n",
      "Sub-losses - total_loss: 61.9073, adver_loss: 17.3972, aggregated_sub_losses: 29.1042, paired_loss: 10.7821, class_loss: 4.6238, multi_bulk.rna.recon_loss: 0.0760, multi_bulk.rna.var_loss: 0.0001, multi_bulk.rna.loss: 0.0760, img.img.recon_loss: 29.0158, img.img.var_loss: 0.0123, img.img.loss: 29.0281, clf_loss: 0.0007\n",
      "split: valid, n_samples: 1681\n",
      "Epoch 8/30 - Valid Loss: 108534.7405\n",
      "Sub-losses - total_loss: 64.5656, adver_loss: 18.9126, aggregated_sub_losses: 30.2427, paired_loss: 10.8168, class_loss: 4.5935, multi_bulk.rna.recon_loss: 0.0779, multi_bulk.rna.var_loss: 0.0000, multi_bulk.rna.loss: 0.0779, img.img.recon_loss: 30.1526, img.img.var_loss: 0.0123, img.img.loss: 30.1648, clf_loss: 0.0002\n",
      "--- Epoch 9/30 ---\n",
      "split: train, n_samples: 11530\n",
      "Epoch 9/30 - Train Loss: 61.9308\n",
      "Sub-losses - total_loss: 61.9308, adver_loss: 17.4222, aggregated_sub_losses: 29.1036, paired_loss: 10.7834, class_loss: 4.6216, multi_bulk.rna.recon_loss: 0.0762, multi_bulk.rna.var_loss: 0.0001, multi_bulk.rna.loss: 0.0763, img.img.recon_loss: 29.0034, img.img.var_loss: 0.0238, img.img.loss: 29.0272, clf_loss: 0.0005\n",
      "split: valid, n_samples: 1681\n",
      "Epoch 9/30 - Valid Loss: 111803.1753\n",
      "Sub-losses - total_loss: 66.5099, adver_loss: 19.5060, aggregated_sub_losses: 31.3465, paired_loss: 11.0013, class_loss: 4.6562, multi_bulk.rna.recon_loss: 0.0779, multi_bulk.rna.var_loss: 0.0001, multi_bulk.rna.loss: 0.0780, img.img.recon_loss: 31.2437, img.img.var_loss: 0.0248, img.img.loss: 31.2685, clf_loss: 0.0011\n",
      "--- Epoch 10/30 ---\n",
      "split: train, n_samples: 11530\n",
      "Epoch 10/30 - Train Loss: 63.0562\n",
      "Sub-losses - total_loss: 63.0562, adver_loss: 18.6493, aggregated_sub_losses: 29.0111, paired_loss: 10.7705, class_loss: 4.6253, multi_bulk.rna.recon_loss: 0.0753, multi_bulk.rna.var_loss: 0.0002, multi_bulk.rna.loss: 0.0755, img.img.recon_loss: 28.8895, img.img.var_loss: 0.0460, img.img.loss: 28.9355, clf_loss: 0.0006\n",
      "split: valid, n_samples: 1681\n",
      "Epoch 10/30 - Valid Loss: 107897.4407\n",
      "Sub-losses - total_loss: 64.1865, adver_loss: 18.7266, aggregated_sub_losses: 30.3607, paired_loss: 10.5741, class_loss: 4.5250, multi_bulk.rna.recon_loss: 0.0774, multi_bulk.rna.var_loss: 0.0002, multi_bulk.rna.loss: 0.0776, img.img.recon_loss: 30.2391, img.img.var_loss: 0.0441, img.img.loss: 30.2831, clf_loss: 0.0000\n",
      "Storing checkpoint for epoch 9...\n",
      "--- Epoch 11/30 ---\n",
      "split: train, n_samples: 11530\n",
      "Epoch 11/30 - Train Loss: 63.3966\n",
      "Sub-losses - total_loss: 63.3966, adver_loss: 18.9577, aggregated_sub_losses: 29.0685, paired_loss: 10.7559, class_loss: 4.6145, multi_bulk.rna.recon_loss: 0.0766, multi_bulk.rna.var_loss: 0.0004, multi_bulk.rna.loss: 0.0770, img.img.recon_loss: 28.9034, img.img.var_loss: 0.0881, img.img.loss: 28.9915, clf_loss: 0.0015\n",
      "split: valid, n_samples: 1681\n",
      "Epoch 11/30 - Valid Loss: 107936.1897\n",
      "Sub-losses - total_loss: 64.2095, adver_loss: 18.2460, aggregated_sub_losses: 30.7006, paired_loss: 10.7021, class_loss: 4.5608, multi_bulk.rna.recon_loss: 0.0753, multi_bulk.rna.var_loss: 0.0004, multi_bulk.rna.loss: 0.0757, img.img.recon_loss: 30.5386, img.img.var_loss: 0.0863, img.img.loss: 30.6249, clf_loss: 0.0000\n",
      "--- Epoch 12/30 ---\n",
      "split: train, n_samples: 11530\n",
      "Epoch 12/30 - Train Loss: 63.0860\n",
      "Sub-losses - total_loss: 63.0860, adver_loss: 18.4938, aggregated_sub_losses: 29.2057, paired_loss: 10.7659, class_loss: 4.6206, multi_bulk.rna.recon_loss: 0.0772, multi_bulk.rna.var_loss: 0.0008, multi_bulk.rna.loss: 0.0780, img.img.recon_loss: 28.9616, img.img.var_loss: 0.1661, img.img.loss: 29.1277, clf_loss: 0.0011\n",
      "split: valid, n_samples: 1681\n",
      "Epoch 12/30 - Valid Loss: 108331.3821\n",
      "Sub-losses - total_loss: 64.4446, adver_loss: 18.5565, aggregated_sub_losses: 30.6605, paired_loss: 10.6752, class_loss: 4.5524, multi_bulk.rna.recon_loss: 0.0771, multi_bulk.rna.var_loss: 0.0005, multi_bulk.rna.loss: 0.0776, img.img.recon_loss: 30.4213, img.img.var_loss: 0.1616, img.img.loss: 30.5829, clf_loss: 0.0000\n",
      "--- Epoch 13/30 ---\n",
      "split: train, n_samples: 11530\n",
      "Epoch 13/30 - Train Loss: 64.3054\n",
      "Sub-losses - total_loss: 64.3054, adver_loss: 19.5508, aggregated_sub_losses: 29.3563, paired_loss: 10.7756, class_loss: 4.6228, multi_bulk.rna.recon_loss: 0.0762, multi_bulk.rna.var_loss: 0.0014, multi_bulk.rna.loss: 0.0777, img.img.recon_loss: 28.9738, img.img.var_loss: 0.3048, img.img.loss: 29.2786, clf_loss: 0.0003\n",
      "split: valid, n_samples: 1681\n",
      "Epoch 13/30 - Valid Loss: 111437.7175\n",
      "Sub-losses - total_loss: 66.2925, adver_loss: 19.9409, aggregated_sub_losses: 30.9175, paired_loss: 10.8357, class_loss: 4.5984, multi_bulk.rna.recon_loss: 0.0762, multi_bulk.rna.var_loss: 0.0014, multi_bulk.rna.loss: 0.0776, img.img.recon_loss: 30.5371, img.img.var_loss: 0.3028, img.img.loss: 30.8399, clf_loss: 0.0000\n",
      "--- Epoch 14/30 ---\n",
      "split: train, n_samples: 11530\n",
      "Epoch 14/30 - Train Loss: 65.7855\n",
      "Sub-losses - total_loss: 65.7855, adver_loss: 20.7457, aggregated_sub_losses: 29.6302, paired_loss: 10.7824, class_loss: 4.6273, multi_bulk.rna.recon_loss: 0.0753, multi_bulk.rna.var_loss: 0.0025, multi_bulk.rna.loss: 0.0778, img.img.recon_loss: 29.0194, img.img.var_loss: 0.5331, img.img.loss: 29.5525, clf_loss: 0.0002\n",
      "split: valid, n_samples: 1681\n",
      "Epoch 14/30 - Valid Loss: 110987.5933\n",
      "Sub-losses - total_loss: 66.0247, adver_loss: 20.2342, aggregated_sub_losses: 30.6272, paired_loss: 10.6313, class_loss: 4.5321, multi_bulk.rna.recon_loss: 0.0778, multi_bulk.rna.var_loss: 0.0020, multi_bulk.rna.loss: 0.0798, img.img.recon_loss: 30.0346, img.img.var_loss: 0.5128, img.img.loss: 30.5474, clf_loss: 0.0000\n",
      "--- Epoch 15/30 ---\n",
      "split: train, n_samples: 11530\n",
      "Epoch 15/30 - Train Loss: 66.3344\n",
      "Sub-losses - total_loss: 66.3344, adver_loss: 21.1345, aggregated_sub_losses: 29.8350, paired_loss: 10.7552, class_loss: 4.6097, multi_bulk.rna.recon_loss: 0.0765, multi_bulk.rna.var_loss: 0.0041, multi_bulk.rna.loss: 0.0805, img.img.recon_loss: 28.8874, img.img.var_loss: 0.8671, img.img.loss: 29.7545, clf_loss: 0.0008\n",
      "split: valid, n_samples: 1681\n",
      "Epoch 15/30 - Valid Loss: 113015.6943\n",
      "Sub-losses - total_loss: 67.2312, adver_loss: 21.2606, aggregated_sub_losses: 30.9571, paired_loss: 10.5173, class_loss: 4.4962, multi_bulk.rna.recon_loss: 0.0781, multi_bulk.rna.var_loss: 0.0023, multi_bulk.rna.loss: 0.0805, img.img.recon_loss: 30.0606, img.img.var_loss: 0.8160, img.img.loss: 30.8767, clf_loss: 0.0000\n",
      "Storing checkpoint for epoch 14...\n",
      "--- Epoch 16/30 ---\n",
      "split: train, n_samples: 11530\n",
      "Epoch 16/30 - Train Loss: 66.8057\n",
      "Sub-losses - total_loss: 66.8057, adver_loss: 21.0939, aggregated_sub_losses: 30.3317, paired_loss: 10.7577, class_loss: 4.6224, multi_bulk.rna.recon_loss: 0.0759, multi_bulk.rna.var_loss: 0.0059, multi_bulk.rna.loss: 0.0818, img.img.recon_loss: 28.9720, img.img.var_loss: 1.2779, img.img.loss: 30.2499, clf_loss: 0.0002\n",
      "split: valid, n_samples: 1681\n",
      "Epoch 16/30 - Valid Loss: 114400.5193\n",
      "Sub-losses - total_loss: 68.0550, adver_loss: 20.5631, aggregated_sub_losses: 32.2344, paired_loss: 10.6868, class_loss: 4.5707, multi_bulk.rna.recon_loss: 0.0779, multi_bulk.rna.var_loss: 0.0054, multi_bulk.rna.loss: 0.0833, img.img.recon_loss: 30.9118, img.img.var_loss: 1.2393, img.img.loss: 32.1512, clf_loss: 0.0000\n",
      "--- Epoch 17/30 ---\n",
      "split: train, n_samples: 11530\n",
      "Epoch 17/30 - Train Loss: 67.0701\n",
      "Sub-losses - total_loss: 67.0701, adver_loss: 20.8530, aggregated_sub_losses: 30.8379, paired_loss: 10.7619, class_loss: 4.6173, multi_bulk.rna.recon_loss: 0.0764, multi_bulk.rna.var_loss: 0.0079, multi_bulk.rna.loss: 0.0844, img.img.recon_loss: 29.0640, img.img.var_loss: 1.6895, img.img.loss: 30.7535, clf_loss: 0.0001\n",
      "split: valid, n_samples: 1681\n",
      "Epoch 17/30 - Valid Loss: 115675.6055\n",
      "Sub-losses - total_loss: 68.8136, adver_loss: 21.1177, aggregated_sub_losses: 32.2924, paired_loss: 10.8021, class_loss: 4.6013, multi_bulk.rna.recon_loss: 0.0763, multi_bulk.rna.var_loss: 0.0062, multi_bulk.rna.loss: 0.0825, img.img.recon_loss: 30.5252, img.img.var_loss: 1.6847, img.img.loss: 32.2099, clf_loss: 0.0000\n",
      "--- Epoch 18/30 ---\n",
      "split: train, n_samples: 11530\n",
      "Epoch 18/30 - Train Loss: 67.5432\n",
      "Sub-losses - total_loss: 67.5432, adver_loss: 21.2476, aggregated_sub_losses: 30.9054, paired_loss: 10.7675, class_loss: 4.6227, multi_bulk.rna.recon_loss: 0.0773, multi_bulk.rna.var_loss: 0.0094, multi_bulk.rna.loss: 0.0867, img.img.recon_loss: 28.7947, img.img.var_loss: 2.0240, img.img.loss: 30.8187, clf_loss: 0.0007\n",
      "split: valid, n_samples: 1681\n",
      "Epoch 18/30 - Valid Loss: 117215.5061\n",
      "Sub-losses - total_loss: 69.7296, adver_loss: 22.1358, aggregated_sub_losses: 32.1851, paired_loss: 10.8079, class_loss: 4.6008, multi_bulk.rna.recon_loss: 0.0778, multi_bulk.rna.var_loss: 0.0069, multi_bulk.rna.loss: 0.0847, img.img.recon_loss: 30.0948, img.img.var_loss: 2.0056, img.img.loss: 32.1004, clf_loss: 0.0000\n",
      "--- Epoch 19/30 ---\n",
      "split: train, n_samples: 11530\n",
      "Epoch 19/30 - Train Loss: 68.6805\n",
      "Sub-losses - total_loss: 68.6805, adver_loss: 21.9283, aggregated_sub_losses: 31.3622, paired_loss: 10.7685, class_loss: 4.6215, multi_bulk.rna.recon_loss: 0.0762, multi_bulk.rna.var_loss: 0.0106, multi_bulk.rna.loss: 0.0868, img.img.recon_loss: 29.0230, img.img.var_loss: 2.2525, img.img.loss: 31.2755, clf_loss: 0.0008\n",
      "split: valid, n_samples: 1681\n",
      "Epoch 19/30 - Valid Loss: 118049.7454\n",
      "Sub-losses - total_loss: 70.2259, adver_loss: 22.3282, aggregated_sub_losses: 32.7928, paired_loss: 10.5867, class_loss: 4.5183, multi_bulk.rna.recon_loss: 0.0800, multi_bulk.rna.var_loss: 0.0086, multi_bulk.rna.loss: 0.0886, img.img.recon_loss: 30.5676, img.img.var_loss: 2.1366, img.img.loss: 32.7042, clf_loss: 0.0000\n",
      "--- Epoch 20/30 ---\n",
      "split: train, n_samples: 11530\n",
      "Epoch 20/30 - Train Loss: 69.0172\n",
      "Sub-losses - total_loss: 69.0172, adver_loss: 22.3097, aggregated_sub_losses: 31.3277, paired_loss: 10.7597, class_loss: 4.6202, multi_bulk.rna.recon_loss: 0.0759, multi_bulk.rna.var_loss: 0.0110, multi_bulk.rna.loss: 0.0869, img.img.recon_loss: 28.8501, img.img.var_loss: 2.3907, img.img.loss: 31.2408, clf_loss: 0.0007\n",
      "split: valid, n_samples: 1681\n",
      "Epoch 20/30 - Valid Loss: 116882.9683\n",
      "Sub-losses - total_loss: 69.5318, adver_loss: 21.6543, aggregated_sub_losses: 32.7290, paired_loss: 10.6148, class_loss: 4.5336, multi_bulk.rna.recon_loss: 0.0778, multi_bulk.rna.var_loss: 0.0095, multi_bulk.rna.loss: 0.0873, img.img.recon_loss: 30.3783, img.img.var_loss: 2.2634, img.img.loss: 32.6417, clf_loss: 0.0000\n",
      "Storing checkpoint for epoch 19...\n",
      "--- Epoch 21/30 ---\n",
      "split: train, n_samples: 11530\n",
      "Epoch 21/30 - Train Loss: 70.1125\n",
      "Sub-losses - total_loss: 70.1125, adver_loss: 23.1954, aggregated_sub_losses: 31.4914, paired_loss: 10.7971, class_loss: 4.6287, multi_bulk.rna.recon_loss: 0.0755, multi_bulk.rna.var_loss: 0.0117, multi_bulk.rna.loss: 0.0872, img.img.recon_loss: 28.9339, img.img.var_loss: 2.4702, img.img.loss: 31.4041, clf_loss: 0.0002\n",
      "split: valid, n_samples: 1681\n",
      "Epoch 21/30 - Valid Loss: 122502.4307\n",
      "Sub-losses - total_loss: 72.8747, adver_loss: 24.7307, aggregated_sub_losses: 32.9238, paired_loss: 10.6650, class_loss: 4.5552, multi_bulk.rna.recon_loss: 0.0778, multi_bulk.rna.var_loss: 0.0067, multi_bulk.rna.loss: 0.0845, img.img.recon_loss: 30.4526, img.img.var_loss: 2.3868, img.img.loss: 32.8394, clf_loss: 0.0002\n",
      "--- Epoch 22/30 ---\n",
      "split: train, n_samples: 11530\n",
      "Epoch 22/30 - Train Loss: 70.7635\n",
      "Sub-losses - total_loss: 70.7635, adver_loss: 23.8077, aggregated_sub_losses: 31.5670, paired_loss: 10.7698, class_loss: 4.6190, multi_bulk.rna.recon_loss: 0.0773, multi_bulk.rna.var_loss: 0.0113, multi_bulk.rna.loss: 0.0886, img.img.recon_loss: 28.9679, img.img.var_loss: 2.5105, img.img.loss: 31.4784, clf_loss: 0.0002\n",
      "split: valid, n_samples: 1681\n",
      "Epoch 22/30 - Valid Loss: 123847.9409\n",
      "Sub-losses - total_loss: 73.6752, adver_loss: 24.9600, aggregated_sub_losses: 33.1439, paired_loss: 10.9306, class_loss: 4.6407, multi_bulk.rna.recon_loss: 0.0789, multi_bulk.rna.var_loss: 0.0093, multi_bulk.rna.loss: 0.0881, img.img.recon_loss: 30.4952, img.img.var_loss: 2.5606, img.img.loss: 33.0558, clf_loss: 0.0000\n",
      "--- Epoch 23/30 ---\n",
      "split: train, n_samples: 11530\n",
      "Epoch 23/30 - Train Loss: 70.9596\n",
      "Sub-losses - total_loss: 70.9596, adver_loss: 24.0684, aggregated_sub_losses: 31.5166, paired_loss: 10.7606, class_loss: 4.6140, multi_bulk.rna.recon_loss: 0.0764, multi_bulk.rna.var_loss: 0.0120, multi_bulk.rna.loss: 0.0884, img.img.recon_loss: 28.8941, img.img.var_loss: 2.5341, img.img.loss: 31.4282, clf_loss: 0.0001\n",
      "split: valid, n_samples: 1681\n",
      "Epoch 23/30 - Valid Loss: 119503.2786\n",
      "Sub-losses - total_loss: 71.0906, adver_loss: 22.7520, aggregated_sub_losses: 33.0421, paired_loss: 10.7291, class_loss: 4.5674, multi_bulk.rna.recon_loss: 0.0769, multi_bulk.rna.var_loss: 0.0077, multi_bulk.rna.loss: 0.0847, img.img.recon_loss: 30.4505, img.img.var_loss: 2.5070, img.img.loss: 32.9575, clf_loss: 0.0000\n",
      "--- Epoch 24/30 ---\n",
      "split: train, n_samples: 11530\n",
      "Epoch 24/30 - Train Loss: 71.6110\n",
      "Sub-losses - total_loss: 71.6110, adver_loss: 24.7039, aggregated_sub_losses: 31.5100, paired_loss: 10.7766, class_loss: 4.6205, multi_bulk.rna.recon_loss: 0.0757, multi_bulk.rna.var_loss: 0.0118, multi_bulk.rna.loss: 0.0875, img.img.recon_loss: 28.8766, img.img.var_loss: 2.5459, img.img.loss: 31.4225, clf_loss: 0.0015\n",
      "split: valid, n_samples: 1681\n",
      "Epoch 24/30 - Valid Loss: 117526.9155\n",
      "Sub-losses - total_loss: 69.9149, adver_loss: 22.1361, aggregated_sub_losses: 32.7363, paired_loss: 10.5202, class_loss: 4.5222, multi_bulk.rna.recon_loss: 0.0774, multi_bulk.rna.var_loss: 0.0092, multi_bulk.rna.loss: 0.0867, img.img.recon_loss: 30.2523, img.img.var_loss: 2.3973, img.img.loss: 32.6497, clf_loss: 0.0000\n",
      "--- Epoch 25/30 ---\n",
      "split: train, n_samples: 11530\n",
      "Epoch 25/30 - Train Loss: 70.0288\n",
      "Sub-losses - total_loss: 70.0288, adver_loss: 23.1721, aggregated_sub_losses: 31.4685, paired_loss: 10.7750, class_loss: 4.6132, multi_bulk.rna.recon_loss: 0.0756, multi_bulk.rna.var_loss: 0.0116, multi_bulk.rna.loss: 0.0873, img.img.recon_loss: 28.8309, img.img.var_loss: 2.5503, img.img.loss: 31.3812, clf_loss: 0.0003\n",
      "split: valid, n_samples: 1681\n",
      "Epoch 25/30 - Valid Loss: 122038.5117\n",
      "Sub-losses - total_loss: 72.5988, adver_loss: 24.3000, aggregated_sub_losses: 32.9660, paired_loss: 10.7402, class_loss: 4.5925, multi_bulk.rna.recon_loss: 0.0774, multi_bulk.rna.var_loss: 0.0070, multi_bulk.rna.loss: 0.0844, img.img.recon_loss: 30.3664, img.img.var_loss: 2.5151, img.img.loss: 32.8816, clf_loss: 0.0000\n",
      "Storing checkpoint for epoch 24...\n",
      "--- Epoch 26/30 ---\n",
      "split: train, n_samples: 11530\n",
      "Epoch 26/30 - Train Loss: 69.3247\n",
      "Sub-losses - total_loss: 69.3247, adver_loss: 22.5462, aggregated_sub_losses: 31.3941, paired_loss: 10.7691, class_loss: 4.6153, multi_bulk.rna.recon_loss: 0.0756, multi_bulk.rna.var_loss: 0.0119, multi_bulk.rna.loss: 0.0876, img.img.recon_loss: 28.7537, img.img.var_loss: 2.5529, img.img.loss: 31.3065, clf_loss: 0.0002\n",
      "split: valid, n_samples: 1681\n",
      "Epoch 26/30 - Valid Loss: 117269.0859\n",
      "Sub-losses - total_loss: 69.7615, adver_loss: 21.7877, aggregated_sub_losses: 32.9031, paired_loss: 10.5364, class_loss: 4.5343, multi_bulk.rna.recon_loss: 0.0783, multi_bulk.rna.var_loss: 0.0061, multi_bulk.rna.loss: 0.0844, img.img.recon_loss: 30.3945, img.img.var_loss: 2.4243, img.img.loss: 32.8188, clf_loss: 0.0000\n",
      "--- Epoch 27/30 ---\n",
      "split: train, n_samples: 11530\n",
      "Epoch 27/30 - Train Loss: 69.7744\n",
      "Sub-losses - total_loss: 69.7744, adver_loss: 22.8115, aggregated_sub_losses: 31.5610, paired_loss: 10.7783, class_loss: 4.6236, multi_bulk.rna.recon_loss: 0.0765, multi_bulk.rna.var_loss: 0.0115, multi_bulk.rna.loss: 0.0880, img.img.recon_loss: 28.9186, img.img.var_loss: 2.5543, img.img.loss: 31.4730, clf_loss: 0.0004\n",
      "split: valid, n_samples: 1681\n",
      "Epoch 27/30 - Valid Loss: 116496.4470\n",
      "Sub-losses - total_loss: 69.3019, adver_loss: 21.4676, aggregated_sub_losses: 32.6469, paired_loss: 10.6493, class_loss: 4.5381, multi_bulk.rna.recon_loss: 0.0777, multi_bulk.rna.var_loss: 0.0062, multi_bulk.rna.loss: 0.0838, img.img.recon_loss: 30.0875, img.img.var_loss: 2.4755, img.img.loss: 32.5630, clf_loss: 0.0000\n",
      "--- Epoch 28/30 ---\n",
      "split: train, n_samples: 11530\n",
      "Epoch 28/30 - Train Loss: 69.0452\n",
      "Sub-losses - total_loss: 69.0452, adver_loss: 22.1127, aggregated_sub_losses: 31.5507, paired_loss: 10.7663, class_loss: 4.6155, multi_bulk.rna.recon_loss: 0.0757, multi_bulk.rna.var_loss: 0.0121, multi_bulk.rna.loss: 0.0878, img.img.recon_loss: 28.9056, img.img.var_loss: 2.5573, img.img.loss: 31.4630, clf_loss: 0.0001\n",
      "split: valid, n_samples: 1681\n",
      "Epoch 28/30 - Valid Loss: 116828.7681\n",
      "Sub-losses - total_loss: 69.4996, adver_loss: 21.7378, aggregated_sub_losses: 32.6463, paired_loss: 10.5662, class_loss: 4.5493, multi_bulk.rna.recon_loss: 0.0760, multi_bulk.rna.var_loss: 0.0071, multi_bulk.rna.loss: 0.0832, img.img.recon_loss: 30.1437, img.img.var_loss: 2.4195, img.img.loss: 32.5631, clf_loss: 0.0000\n",
      "--- Epoch 29/30 ---\n",
      "split: train, n_samples: 11530\n",
      "Epoch 29/30 - Train Loss: 69.1178\n",
      "Sub-losses - total_loss: 69.1178, adver_loss: 22.2346, aggregated_sub_losses: 31.4871, paired_loss: 10.7742, class_loss: 4.6219, multi_bulk.rna.recon_loss: 0.0757, multi_bulk.rna.var_loss: 0.0121, multi_bulk.rna.loss: 0.0877, img.img.recon_loss: 28.8419, img.img.var_loss: 2.5575, img.img.loss: 31.3993, clf_loss: 0.0000\n",
      "split: valid, n_samples: 1681\n",
      "Epoch 29/30 - Valid Loss: 117392.3801\n",
      "Sub-losses - total_loss: 69.8348, adver_loss: 21.6843, aggregated_sub_losses: 32.7827, paired_loss: 10.7883, class_loss: 4.5795, multi_bulk.rna.recon_loss: 0.0770, multi_bulk.rna.var_loss: 0.0122, multi_bulk.rna.loss: 0.0892, img.img.recon_loss: 30.1630, img.img.var_loss: 2.5305, img.img.loss: 32.6935, clf_loss: 0.0000\n",
      "--- Epoch 30/30 ---\n",
      "split: train, n_samples: 11530\n",
      "Epoch 30/30 - Train Loss: 68.7689\n",
      "Sub-losses - total_loss: 68.7689, adver_loss: 21.9025, aggregated_sub_losses: 31.4829, paired_loss: 10.7602, class_loss: 4.6234, multi_bulk.rna.recon_loss: 0.0749, multi_bulk.rna.var_loss: 0.0118, multi_bulk.rna.loss: 0.0867, img.img.recon_loss: 28.8400, img.img.var_loss: 2.5562, img.img.loss: 31.3962, clf_loss: 0.0000\n",
      "split: valid, n_samples: 1681\n",
      "Epoch 30/30 - Valid Loss: 116418.0056\n",
      "Sub-losses - total_loss: 69.2552, adver_loss: 21.3000, aggregated_sub_losses: 32.9048, paired_loss: 10.5366, class_loss: 4.5138, multi_bulk.rna.recon_loss: 0.0777, multi_bulk.rna.var_loss: 0.0100, multi_bulk.rna.loss: 0.0877, img.img.recon_loss: 30.3851, img.img.var_loss: 2.4320, img.img.loss: 32.8170, clf_loss: 0.0000\n",
      "Storing checkpoint for epoch 29...\n",
      "Prediction complete.\n",
      "Processing latent space results into a single AnnData object...\n",
      "Identified source modality for latent space: 'multi_bulk.rna'\n",
      "  - Added 20 source feature IDs to .uns\n",
      "Finished processing latent results.\n"
     ]
    }
   ],
   "source": [
    "rna_file = os.path.join(\"data/XModalix-Tut-data/combined_rnaseq_formatted.parquet\")\n",
    "img_root = os.path.join(\"data/XModalix-Tut-data/images/tcga_fake\")\n",
    "\n",
    "xmodalix_config = XModalixConfig(\n",
    "    checkpoint_interval=5,\n",
    "    class_param=\"CANCER_TYPE_ACRONYM\",\n",
    "    epochs=30,\n",
    "    beta=0.01,\n",
    "    gamma=2,\n",
    "    delta_class=2.0,\n",
    "    delta_pair=3.0,\n",
    "    latent_dim=8,\n",
    "    requires_paired=False,\n",
    "    pretrain_epochs=20,\n",
    "    data_case=DataCase.IMG_TO_BULK,\n",
    "    data_config=DataConfig(\n",
    "        data_info={\n",
    "            \"img\": DataInfo(\n",
    "                file_path=img_root,\n",
    "                data_type=\"IMG\",\n",
    "                scaling=\"MINMAX\",\n",
    "                translate_direction=\"to\",\n",
    "                # extra_anno_file=imganno_file,\n",
    "            ),\n",
    "            \"rna\": DataInfo(\n",
    "                file_path=rna_file,\n",
    "                data_type=\"NUMERIC\",\n",
    "                scaling=\"MINMAX\",\n",
    "                translate_direction=\"from\",\n",
    "            ),\n",
    "            \"anno\": DataInfo(file_path=clin_file, data_type=\"ANNOTATION\", sep=\"\\t\"),\n",
    "        },\n",
    "    ),\n",
    ")\n",
    "\n",
    "xmodalix = acx.XModalix(config=xmodalix_config)\n",
    "result = xmodalix.run()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a8b25ea3",
   "metadata": {},
   "outputs": [],
   "source": [
    "xmodalix.predict()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d5cf72ae",
   "metadata": {},
   "source": [
    "## 2) XModalix Specific Steps\n",
    "\n",
    "XModalix does not introduce additional pipeline steps. However, its preprocessing and training differ slightly from Varix and Vanillix. XModalix processes each data modality individually and therefore supports unpaired data input.  \n",
    "\n",
    "It is also possible to re-run the `predict` step and change the translation pairs or direction by providing the `from_key` and `to_key` arguments with the string keys that define your data modalities (e.g., `\"rna\"` and `\"img\"` in our case). See the code below for details.\n",
    "\n",
    "Furthermore, the visualization and evaluation steps produce outputs that differ somewhat from the other pipelines, which we will focus on in [Section 4](#4-visualize-and-evaluate-xmodalix).\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dd55812e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Initially, we translated from RNA to image. Now we can use the trained model\n",
    "# to re-run the translation with switched pairs.\n",
    "# ATTENTION: This will overwrite the existing result object, so we create a backup first.\n",
    "from copy import deepcopy\n",
    "\n",
    "backup_result = deepcopy(result)\n",
    "result_flipped = xmodalix.predict(from_key=\"img\", to_key=\"rna\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "92c18f33",
   "metadata": {},
   "source": [
    "## 3) Access XModalix Results\n",
    "\n",
    "Accessing the results works slightly differently for `XModalix` because we have:\n",
    "\n",
    "- Multiple latent spaces (one for each data modality).\n",
    "\n",
    "- Not just reconstructions, but also `translations` and `reference translations`.  \n",
    "  *Reference translations* are reconstructions within the same data modality as the translation split.  \n",
    "  For example, if we translate from `rna` to `img`, the reference translation is the reconstruction from `img` to `img`.  \n",
    "  These translations can be accessed via the `reconstructions` attribute of the `result` object.  \n",
    "  First, use the usual `TrainingDynamics` API via `.get()`, which returns a dict containing each data modalityâ€™s translation and reference translation.\n",
    "\n",
    "- Multiple losses.  \n",
    "  These are accessed via the `sub_losses` attribute of the `result` object. This is a `Dict` of `TrainingDynamics`.  \n",
    "  First, select the loss type youâ€™re interested in, then work with the standard `TrainingDynamics` API.\n",
    "\n",
    "See the code below for examples and [3] for more details:\n",
    "\n",
    "[3] `Tutorials/DeepDives/PipelineOutputTutorial.ipynb`\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1599265d",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(\"Find available loss types\")\n",
    "print(backup_result.sub_losses.keys())\n",
    "print(\"Get adver_loss for training\")\n",
    "backup_result.sub_losses.get(\"adver_loss\").get(split=\"train\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b822c090",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(\"Get reconstruction keys\")\n",
    "# Frist define split and epoch you're interested in\n",
    "# usually test split (there are no epochs, so by default this is always epoch=-1)\n",
    "recons = backup_result.reconstructions.get(split=\"test\", epoch=-1)\n",
    "print(recons.keys())\n",
    "print(\"Getting Translation\")\n",
    "trans = recons.get(\"translation\")\n",
    "print(f\"shape of translation: {trans.shape}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "161f5101",
   "metadata": {},
   "source": [
    "**We can acutally plot a sample image here**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1ce547d2",
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "\n",
    "sample = trans[0, 0, :, :]\n",
    "plt.imshow(sample)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "84d69d35",
   "metadata": {},
   "source": [
    "Accessing the latent spaces works analogously."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e3783d19",
   "metadata": {},
   "source": [
    "## 4) Visualize and Evaluate XModalix\n",
    "\n",
    "XModalix offers a few additional visualizations, namely:\n",
    "\n",
    "- An image grid (when using images) that compares original with translated data.  \n",
    "- A visualization of each data modality's latent space, which can be colored based on a target variable.\n",
    "\n",
    "Additionally, the `evaluate` step works differently for XModalix. While for other pipelines we evaluate embedding quality with downstream machine learning tasks, here we focus on how well biological information is preserved in the translated data modality. We typically plot the mean squared error (MSE) between the original data and the translated data, clustered by a metadata parameter (e.g., cancer type).\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b0c089fa",
   "metadata": {},
   "outputs": [],
   "source": [
    "fig = xmodalix._visualizer.show_2D_translation(\n",
    "\tresult=backup_result,\n",
    "\ttranslated_modality=\"img.img\",\n",
    "\tsplit=\"test\",\n",
    "\tparam=\"CANCER_TYPE_ACRONYM\",\n",
    "\treducer=\"UMAP\",\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "eb203808",
   "metadata": {},
   "source": [
    "### ATTENTION\n",
    "Eval needs fixing from JE"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bb5cc0a2",
   "metadata": {},
   "outputs": [],
   "source": [
    "xmodalix.evaluate()\n",
    "param = \"early\"\n",
    "fig, df_comparison = xmodalix._evaluator.pure_vae_comparison(\n",
    "\txmodalix_result = xmodalix.result,\n",
    "\t# pure_vae_result = imagix.result,\n",
    "\tto_key = \"img.IMG\",\n",
    "\tparam = param\n",
    ")\n",
    "\n",
    "xmodalix._visualizer.plots[\"PureVAEComparison\"][param] = fig\n",
    "fig.figure\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d02aef37",
   "metadata": {},
   "source": [
    "## 5) Apply Custom Parameters\n",
    "Due to the more complex training, we have more config parameters to customize. Here we go over the different `loss terms` and `pretraining`.\n",
    "```python\n",
    "    gamma: float = Field(\n",
    "        default=10.0,\n",
    "        ge=0,\n",
    "        description=\"Gamma weighting factor for Adversial Loss Term i.e. for XModal Classfier training\",\n",
    "    )\n",
    "    delta_pair: float = Field(\n",
    "        default=5.0,\n",
    "        ge=0,\n",
    "        description=\"Delta weighting factor for paired loss term in XModale Training\",\n",
    "    )\n",
    "    delta_class: float = Field(\n",
    "        default=5.0,\n",
    "        ge=0,\n",
    "        description=\"Delta weighting factor for class loss term in XModale Training\",\n",
    "    )\n",
    "```\n",
    "The correct parametrization of these terms depends heavily on your data. We provide sensible default values in our `XModalixConfig` class, which are a good starting point. You can investigate the loss plots to see how much each loss contributes. For example, if your goal is to strongly preserve class information in the latent embedding, you would set `delta_class` higher. If you care more about generalization, you might set `gamma` higher.\n",
    "\n",
    "Another important factor is pretraining. Before training a modality within the full XModalix architecture, we can pretrain that modality in its subnetwork without the XModalix constraints (adversarial, paired, class loss). This can be controlled via the `pretrain_epochs` parameter, which can be set either globally for all data modalities or individually per modality. Both options are shown below.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "80f47d10",
   "metadata": {},
   "outputs": [],
   "source": [
    "import autoencodix as acx\n",
    "from autoencodix.configs.xmodalix_config import XModalixConfig\n",
    "from autoencodix.configs.default_config import DataConfig, DataInfo, DataCase\n",
    "\n",
    "\n",
    "xmodalix_config = XModalixConfig(\n",
    "    class_param=\"CANCER_TYPE_ACRONYM\",\n",
    "    gamma=2,\n",
    "    delta_class=10.0,  # increase this if you care about class info in embedding\n",
    "    delta_pair=3.0,\n",
    "    requires_paired=False,\n",
    "    pretrain_epochs=2,  # can be overridden on data level, see below\n",
    "    data_case=DataCase.IMG_TO_BULK,\n",
    "    data_config=DataConfig(\n",
    "        data_info={\n",
    "            \"img\": DataInfo(\n",
    "                file_path=\"YOUR_PATH\",\n",
    "                data_type=\"IMG\",\n",
    "                translate_direction=\"to\",\n",
    "                pretrain_epochs=2,  # this overrides global pretraining_epochs\n",
    "            ),\n",
    "            \"rna\": DataInfo(\n",
    "                file_path=\"YOUR_PATH\",\n",
    "                data_type=\"NUMERIC\",\n",
    "                translate_direction=\"from\",\n",
    "            ),\n",
    "            \"anno\": DataInfo(file_path=\"YOUR_PATH\", data_type=\"ANNOTATION\", sep=\"\\t\"),\n",
    "        },\n",
    "    ),\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7d665447",
   "metadata": {},
   "source": [
    "## 6) Save, Load and Reuse XModalix\n",
    "\n",
    "This works for `XModalix` as for any other models, by using the `save` and `load` methods."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ba9b7961",
   "metadata": {},
   "outputs": [],
   "source": [
    "xmodalix_loaded._trainer._fabric.device"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "39525727",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "from lightning_fabric.wrappers import _FabricModule\n",
    "models = xmodalix_loaded.result.model\n",
    "keys = models.keys()\n",
    "print(keys)  # dict_keys(['img_encoder', 'img_decoder', 'rna_encoder',\n",
    "img_model = models[\"img.img\"]\n",
    "rna_model = models[\"multi_bulk.rna\"]   \n",
    "fabric = xmodalix_loaded._trainer._fabric\n",
    "print(f\"type of img_model: {type(img_model)}\")\n",
    "print(f\"type of rna_model: {type(rna_model)}\")\n",
    "print(isinstance(rna_model, _FabricModule))\n",
    "img_model_setup = fabric.setup_module(img_model)\n",
    "rna_model.to(fabric.device)\n",
    "\n",
    "img_model.to(fabric.device)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3573f6c4",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import glob\n",
    "\n",
    "outpath = os.path.join(\"xmodalix\")\n",
    "xmodalix.save(file_path=outpath, save_all=False)\n",
    "\n",
    "folder = os.path.dirname(outpath)\n",
    "pkl_files = glob.glob(os.path.join(folder, \"*.pkl\"))\n",
    "model_files = glob.glob(os.path.join(folder, \"*.pth\"))\n",
    "\n",
    "print(\"PKL files:\", pkl_files)\n",
    "print(\"Model files:\", model_files)\n",
    "\n",
    "# the load functionality automatically will build the pipeline object out of the three saved files\n",
    "xmodalix_loaded = acx.XModalix.load(outpath)\n",
    "# now you can use the model to predict with a different pair again:\n",
    "testdata = backup_result.datasets\n",
    "r = xmodalix_loaded.predict(data=testdata)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "92d8b378",
   "metadata": {},
   "outputs": [],
   "source": [
    "xmodalix_loaded.result.model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fc36a328",
   "metadata": {},
   "outputs": [],
   "source": [
    "xmodalix_loaded.show_result()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "autoencodix",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
